{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "96347609",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "# Import libraries\n",
    "%pylab inline\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a3295ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load sample order and labels\n",
    "sampleinfo = pd.read_csv(\"~/public/1000Genomes/igsr_samples.tsv\", sep=\"\\t\")\n",
    "sample_to_pop = dict(zip(list(sampleinfo[\"Sample name\"]), list(sampleinfo[\"Population code\"])))\n",
    "samples = [line.split()[0] for line in open(\"ps2_admixture.pruned.fam\", \"r\").readlines()]\n",
    "pops = [sample_to_pop.get(item, \"NA\") for item in samples]\n",
    "for item in pops:\n",
    "    assert(item in [\"CEU\", \"PEL\", \"GWD\", \"ASW\", \"PUR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a491af11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n",
      "These are the actual results from the admixture tool \n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAAGRCAYAAAB4wkXUAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAAsTAAALEwEAmpwYAAAzpklEQVR4nO3de7xcdX3v/9fHBCTFIHK/BE4CBCUCCkYU8KEYpCehrYC0aaB4l0gNLXgn6gOV0yqKrVJFMVhv9acR+6uUUwNqD6QIAiUFNUKKJyWaBOQWxUYgkoTP+WNmh8lmX2bvPbPWrFmv5+OxH8y67LU+ZM+e/Z7vfNZ3RWYiSZIkqeEZZRcgSZIk9RIDsiRJktTCgCxJkiS1MCBLkiRJLQzIkiRJUovJZZ147ty5ee2115Z1ekmSJCmGWlnaCPLDDz9c1qklSZKkYdliIUmSJLUwIEuSJEktDMiSJElSi1Ev0ouILwJ/CDyYmYcPsT2AS4GTgceAN2Tm7eMpZvPmzaxfv55NmzaN59sLsdNOOzFt2jR22GGHskuRJElSF7Qzi8WXgc8AXx1m+zxgZvPrJcDnmv8ds/Xr1zN16lSmT59OI3f3lsxkw4YNrF+/nhkzZpRdjiRJkrpg1BaLzLwB+NUIu5wCfDUbbgF2jYh9x1PMpk2b2H333XsyHANEBLvvvntPj3BLkiRpYjoxD/L+wLqW5fXNdb8cvGNELAQWAuyw+w4c8ZUjttv+qVmf4skNT3agpO66/7f3M/8r88suY0Qr16wF4IgZB257LI3VETMOLLsEaUitr2sDz9PR1rUa6bWxdVvr78BEX0tHO2cnztGu6Zu+Xsh5JmrqYRd09fij/XsP9Rq4cs3aYX9eo+0/lrom8j3tPO5ELUP9zg3eNtzv0HC/Z0Odc7hjDffvPdQxh9q+8vUrh9wHCr5RSGYuAZYATJkxJYs8tyRJ0ng52FQvnQjI9wIHtCxPa66bsD/4xM87cZhtvvOu6W3td+P/uZGL338xW7du5fSzTuct572lo3VIklSWqoweS2XqxDRvVwOvi4aXAr/JzKe1V1TF1q1b+asL/orPLf0cV990Ncu+vYz/uvu/yi5r3HzHK0mSqmpwjlm5Zm0h2aadad6+AZwA7BER64EPAjsAZOblwDIaU7ytpjHN2xu7VWwRVt6+kgOnH8gB0xuD4vNOncd111zHwc89uOTKJElSJ/TS4NFoPbMqx6gBOTPPGGV7Aos6VlHJHvzlg+yz/z7blvfeb29W/sfwTdySJPU62yomppcCdRnKCPFl/5t7Jz1JkqQ+MNb2g7JDaFHG05ZR6CwWVbDXvntx/733b1t+4L4H2GvfvUqsSJI0oC5/0NU9PofaU/d/JwPyIIcfdThr16xl/S/Ws/e+e3PNVdfw8cs/XnZZkiSNiW0VKks74brXA3hPB+R2p2XrpMmTJ/O+j76Pt85/K1uf3MppZ5zGIc87pPA6JEkaD4OxumU8Nxrppm72Rvd0QC7Ly096OS8/6eVllyFJ0pgYjqXO8CI9SVLl9frHtaqHdi8GG8tFYz63x2ei8yUbkCVJktSXxhuSbbGQJEl9z5FYjYUBWZKkirHXWBNVxTcMRdZsQJYkSeoBowXAoWaRqGLQrQJ7kCVJUl8zRGqsenoE+fmfOa6jx7vz3B+Ous8H/vID3PD9G9htj9246gdXdfT8kqTeYGCqh7J+zr02X3C7ujmvcNU4gjzIqQtO5fKll5ddhiSpB/RSkJ6+6evbviR1lwF5kNnHzebZz3l22WVIkkrSS6FYGsznZzEMyJKkSjMwSOo0A7IkSVJBqv6GbqJ3qKuKnr5IT5IkOe+x6qGXgrcBWZKkHmUwHr9uhq1eCnLD6aUae6mWdvV0QG5nWrZOe/fCd3PbTbfxyK8e4cQjT+Rt73kbp591euF1SJLqzXAstacb0+r1dEAuwyVLLim7BEmSpFF1YmTWuY+H1tZFehExNyLujojVEXHBENsPjIjrI+KOiPhJRJzc+VIlSZL6RxVbD+pi1BHkiJgEXAacBKwHbouIqzPzrpbdPgBcmZmfi4hZwDJgehfqlSSpb9lWMXGGzpH579OedlosjgFWZ+Y9ABGxFDgFaA3ICezSfPxs4L5OFilJkjQaw199dfpn305A3h9Y17K8HnjJoH0+BHwvIv4C2Bl41VAHioiFwEKAHXbfYay1SpLUlzauurjsEirPcKxO6tSNQs4AvpyZ04CTgX+IiKcdOzOXZObszJw9aeqkDp1akqTqsq1C3VbmzT2q+salnRHke4EDWpanNde1ejMwFyAzb46InYA9gAc7UWRdeCWpJNWDoVgq3ljCejsB+TZgZkTMoBGMFwBnDtpnLXAi8OWIOAzYCXio7SqGseA7CyZ6iO0s/YOlo+7zy3t/yfsWvY8ND20gIvjj1/4xr33raztahySpngzGUjWMGpAzc0tEnAt8F5gEfDEz74yIi4AVmXk18E7gioh4O40L9t6QmdnNwrtl8qTJvPvD72bWC2bx6G8fZf6J8znuhOM4+LkHl12aJKlitgvEq8qrQ+q0qrZOtKutG4Vk5jIaU7e1rruw5fFdwPGdLa0ce+6zJ3vusycAOz9rZw469CAe+OUDBmRJklSYfg+gva5TF+n1pXvX3suqlas48kVHll2KJEmSCuKtpofx2G8f4+1vfDvv/av38qypzyq7HEmSNARHWjUeoz1vDMhD2Lx5M+e/8Xz+4I//gJP+8KSyy5EkVZAX5NWbwb3aDMiDZCYXnn8hBx16EK//89eXXY4kaRzKCieGYml8eu0NRU8H5HamZeu0O269g/995f9m5qyZnH7C6QCc9/7zePlJLy+8FklSNRiMNZKVa9Z6j4OK6emAXIajX3o0P33op2WXIUmSpJI4i4UkSZLUwhFkSZLGwbaK/tdrfbEqTk+NICdJr9+ALzNJertGSVJ3GY6l/tZTI8jrHl/Hbht3Y8epOxIRZZfzNJnJExufYN3j68ouRZJUAoOxVA89FZCvWHsFZ3M2B0w5gKAHAzLJusfXccXaK8ouRZJUMMOxVB89FZA3bt3I367527LLkCRJKtRI/c72Qhevp3qQJUmSpLIZkCVJkqQWBmRJkiSphQFZkiRVkr256paeukhPkqRes3HVxWWXIKlgBmRJkgZxSjep3gzIkiSpUmytULfZgyxJkiS1MCBLkiRJLdoKyBExNyLujojVEXHBMPvMj4i7IuLOiLB5S5JUSfYfSxq1BzkiJgGXAScB64HbIuLqzLyrZZ+ZwGLg+Mz8dUTs1a2CJUmSpG5qZwT5GGB1Zt6TmU8AS4FTBu1zNnBZZv4aIDMf7GyZkiR1n6PHvc8L9FSEdmax2B9Y17K8HnjJoH0OBYiIm4BJwIcy89rBB4qIhcBCgB1232E89UqS1HEGY0mtOjXN22RgJnACMA24ISKOyMxHWnfKzCXAEoApM6Zkh86tpoF31UfMOLDkSiRJkqqrnYB8L3BAy/K05rpW64FbM3MzsCYifkYjMN/WkSolSeowR42l+pq+6ev8fITt7QTk24CZETGDRjBeAJw5aJ+rgDOAL0XEHjRaLu4Zc7WSJHWRoVganX3ebQTkzNwSEecC36XRX/zFzLwzIi4CVmTm1c1tvx8RdwFbgXdn5oZuFi5JkqTe0U/Buq0e5MxcBiwbtO7ClscJvKP5JUmSJFWWd9KTJEmquX4a/e0EA7IkSZLUolPTvKnHDFyI8vOdBl9PKUn15AV6ktp9HXAEuUf40YYkSVJvcARZktTXHDmWuq/fBvoMyCXrtyeUJElS1dliUSLDsSRJUu8xIFfAWIP0xlUXd6kSSZKk/mdAliT1LfuPpd7Xi5+o24MsSeo7BmNJE+EIcsWtXLO2J995SVJZDMeSJsqALEmSJLWwxaJg3RrtXblmraMmkmrN10BJnWJALoAtEJIkSdVhQO4ig7EkSVL12IMsSZIktXAEuUscPZak4th/rDoxY3SfI8iSJElSCwOyJEmS1MKALEmSJLUwIEuSKs3+Y0md1tZFehExF7gUmAR8ITMvHma/04F/BF6cmSs6VmVF2DQvSZKGYkaollEDckRMAi4DTgLWA7dFxNWZedeg/aYC5wG3dqNQSZJaOXIsqVvaabE4Blidmfdk5hPAUuCUIfb7X8DHgE0drE+SJEkqVDsBeX9gXcvy+ua6bSLiaOCAzPzOSAeKiIURsSIiVmzduHXMxWp4jqRIkvrZyjVrbVNQYSZ8kV5EPAP4W+Cdo+2bmUsyc3Zmzp40ddJETy1JkiS1ZSyDie0E5HuBA1qWpzXXDZgKHA4sj4ifAy8Fro6I2W1XoVH5zlmSJKkY7cxicRswMyJm0AjGC4AzBzZm5m+APQaWI2I58K66zWJheJUkSeot421BHXUEOTO3AOcC3wVWAVdm5p0RcVFEvHpcZ5UkSZJ6VFvzIGfmMmDZoHUXDrPvCRMvS5IkSRqfiU5e0FZA1vBsrZAkSeov3mpakiRJamFAliRJUt/oxL0hDMgTYHuFJElS/zEgV5gBXZIkqfMMyJIkSVILZ7EYB0duJUmSekcn+o5bGZDHqBfCcS/UIEll6vQfQ0nV083XAQNyGwykkiRJ9WEPsiRJktTCEeQROHIsSVL5/HusojmCLEmSpErp9nUIjiBLkiSpEoq6QNeAXHFeyS1JktRZBmRJkiT1rDIGAw3Ig3ghgCRJUr0ZkCVJktRTym4hdRYLSZIk9YyywzE4grwd2yskSZLK0QvBeIAjyJIkSVILA7IkSZJK1Uujx9Bmi0VEzAUuBSYBX8jMiwdtfwfwFmAL8BDwpsz8RYdrlSRJUh/otUA82KgBOSImAZcBJwHrgdsi4urMvKtltzuA2Zn5WET8OfBx4E+7UbAkSZKqo9fD8FDaGUE+BlidmfcARMRS4BRgW0DOzOtb9r8FOKuTRUqSpPrx4vnqqmIobtVOD/L+wLqW5fXNdcN5M3DNUBsiYmFErIiIFVs3bm2/SkmSJFVC1cMxdHiat4g4C5gNvGKo7Zm5BFgCMGXGlOzkuSfCd6iSJEnj1w+huFU7Afle4ICW5WnNdduJiFcB7wdekZm/60x53Wc4liRJUqt2AvJtwMyImEEjGC8AzmzdISKOAj4PzM3MBztepSRJknpKv40atxq1BzkztwDnAt8FVgFXZuadEXFRRLy6udslwLOAb0XEjyLi6q5VrG36+YkpSZJUlrZ6kDNzGbBs0LoLWx6/qsN1FcL2CkmSpPbVZXCuoxfpqbMM8JIkqWh1CcEjMSBLkiTVnKF4ewZkSZKkmjIYD62WAdnWBUmSVHeG4+HVLiAbjiVJUl0YgsennVtNS5IkSbVRmxHkfho59t2gpLry9U8anr8fnVOLgNxP4XjjqovLLkGSJPUAA3H39G1A7qdQLElVNvB6fMSMA0uuROoPBuPu69uALEmS1A8MxMXri4DsaLEkSeonhuJyVX4WC8OxJEnqJ4bj8vXFCLIkSVKVGYp7iwFZkiSpIAbhajAgS5IkdZnBuFoq34MsSZIkdVKlR5C9QE+SepOvz9JTHD2uHkeQJUkdZTiWVHWVHkGuE999SpJUDf7Nrr5KBuS6jU74iyapCur22iyBf6P7VeUCsi/AUjGmb/o6U7mg7DJUEb42qy4MxPVQuYBcJ/4Sqig+1yRpZL5O1osBWaopX+zVCY4cq5/5OllfkZnlnDjiIeBR4GFgj+bq1seDlyfy2GNVqxaPVd1aPJbPA49V7Vo8VnVr8Vjje/yfmTmXQUqb5i0z9wQezszZNIrc7vFI28b62GNVqxaPVd1aPJbPA49V7Vo8VnVr8Vjje5xDhGNwHmRJkiRpOwZkSZIkqcXkks+/ZNB/Bz8eadtYH3usatXisapbi8cq71i9VIvHqm4tHqu6tXisiR1rm9Iu0pMkSZJ6kS0WkiRJUgsDsiRJktTCgCxJkiS1MCBLkiRJLQoJyBHxqYiYHxH7F3E+SVL/i4hvll2DihERJ0XE98uuQ/VR1DRvq4FTgY9HBMAPm183AT/OzCcLqqOvRcSngdZpSZLGHWOuz8wby6lKVRIRzwUWAs9rrloFXJGZd5dXlTSsY8suQJ0VEXOAy4H9gKuAjwFfAgL46/IqU90UPs1bROwHHNf8ejWwV2buUmgRfSoiXj/E6t2A+cA3M/NTxVakKomIY4F/Aj4P3EHjD9JRwNnAazLzlhLLk54mItZm5oFl16HOiYg7gLcDNwPzgK8BF2TmZ0otTLVTWECOxtDxETSC8fHALOAh4ObM/HAhRdRUREwBfpiZR5Vdi3pXRFwDfCwzlw9a/woaf6DmlVKYai0ijh5uE/AvmblvkfWouyLi9sw8umX57sx8bpk1qZ4KabFo9g3tAvwIuAX4SGauKuLcgsx8vNnaIo3k4MHhGCAz/y0ihr3bkNRlfzPCtv8srAoVZdeIeE3L8uTW5cz8pxJqUg0V1YN8D3AkMBPYADwcEQ9l5sMFnb+2ImIy8Fpgfdm1qOdtHGHbo4VVIbXIzFeWXYMK9W/AH7Us39CynDTawKSuK7QHOSJ2AV5Ko83ipcCewE8zc6jeWY1RRGxk+4v0AB6n8YJzfmbeV3xVqoqIeBBYOtQmYH5m7l1wSRIR8Z7M/Hjz8Z9k5rdatn0kM99XXnWS+lXRAfmZwItp9CAPhOQHM/OIwoqQNKRhLvLcJjO/UlQt0oDWntQh+lO3W1b1RcQ7Bq0amI3pxsxcU0JJqqmiepA/SSMQz6RxdfzNNKZxeX1mPlJEDXUQEWdl5teaj4/PzJtatp3rVcAaxY9pTLtY7NQ20shimMdDLav6pg6xbjrw/oj4UGYO9SmX1HGFjCBHxF/SmPP4R5m5tesnrClHWjQREbECOAj4D56ap/zmzBypN1nqKl/XBBARuwH/6s9bRSnqIr1/Bh4ZCMcR8UoaNw75BfCZzHyioDr6nSMtGrfMnB0RvwccQ+MTn78E/iEi7gduysy3lVqg6uoFEfHfNF7DpjQf01zeqbyyVKTM/FU4HZMKVMitpoErgZ0BIuKFwLeAtcALgM8WVEMdDL6L3nDbpCFl5mPNqd4uBT4JXEbjd3dumXWpvjJzUmbukplTM3Ny8/HA8g5l16diNAfWfl12HaqPokaQp7TMoHAW8MXM/JuIeAaNuZHVGc+LiJ/QGFk5uPmY5vJB5ZWlKoiIM2mMHL8Q+B1wG3Ar8LLMvL/E0lRjEbETcA5wCPATGn8/tpRblbolIlby9AGd3YD7gNcVX5HqqqiA3PqxyBxgMUBmPulHJh31AmBvYN2g9QcABhyN5vPA3TQuoL0hM39Wcj0SwFeAzcAPgJOB5wPnlVqRuukPBy0nsCEznYtdhSoqIF8XEVcCvwSeA1wHEBH7ApsKqqEOPgkszsxftK5szj/9SbaffF0abFcab7KOAz4UEc+l8Tt7M42L9a4rsTbV16yBqUAj4u+Bfy+5HnXXAzz1icFK4O/9xEBlKKoH+Xwad7/5OY2Pazc31x9C46MTdcbembly8MrmuunFl6MqycytmXl7Zn4mM8+kMVp3LfBG4PvlVqcaG/h7gUGpFr4CzKYRjucx8q3Gpa4pZAS5Oa/qUoCIOCoizgf+BFgDfKqIGmpi1xG2TSmqCFVTRBxJY/R44GtHGqPHn6Yx5ZtUhhcMmrliSsusFpmZu5RXmrrATwzUE4q6UcihwBnNr4eBb9KYg/mVRZy/RlZExNmZeUXryoh4C425baWRfBm4EbgG+EBmri23HKkxi0XZNahQ231i4GVKKktRNwp5ksYFFm/OzNXNdfdkpjMrdFBE7A18G3iCpwLxbBojgac5E4FGEhGnAtMG7rgYEbcCezY3vycz/7Gs2iTVQ0RsBQYuyAsan34+hp8YqGBFXaT3GmABcH1EXEuj3cK3hR2WmQ8AxzXnizy8ufo7XlylNr2bxu/pgGcCL6YxD/KXAAOypK7yEwP1iqJ6kK8CroqInYFTaFy0t1dEfA74dmZ+r4g66iIzrweuL7sOVc6Omdk6ReCNmbkB2ND83ZUkqRYKabEY8sQRz6Fxod6fZuaJpRQhaZuIWJ2Zhwyz7b8y8+Cia5IkqQxFTfP2NJn568xcYjiWesatEXH24JUR8Va8klySVCOljSBL6i0RsRdwFY3bTN/eXP0iGr3IpzZ73CVJ6nsGZEnbiYg5NG7nC3CnF3lKkurGgCxJkiS1KK0HWZIkSepFBmRJkiSphQFZkiRJamFAliRJkloYkCVJkqQWBmRJkiSphQFZkiRJajG5rBPPnTs3r7322rJOL0mSJMVQK0sbQX744YfLOrUkSZI0LFssJEmSpBYGZEmSJKmFAVmSJElqMWpAjogvRsSDEfHTYbZHRPxdRKyOiJ9ExNGdL1OSJEkqRjsjyF8G5o6wfR4ws/m1EPjcxMuSJEmSyjFqQM7MG4BfjbDLKcBXs+EWYNeI2LdTBUqSJElF6sQ8yPsD61qW1zfX/XLwjhGxkMYoM/tOnsyq5x02rhNed8JlAMxZvojrTriMOcsXDblP6/rW7xlpn6G+Z/D3Dbc/wGEL7nvafquW7tfWsVq3D1fLon1O47L7v/207we4/NjzAFi5Zu2w+7Sef/7i0X/859x8KZfs+vio+6k/XXPVu7Y9Hvw7NNpzul3zF0/myo9u2fZ45Zq14z6W+tOqpfsBjdfXoV7b5ixf1NbrWVnOuflSoPEaPfBch/Zeg1uPMfAaP9q2gfMNnHMixx7pe1qP3/o73Hr+1joGf89o9bd+z3gN9Xo10r7t7DeRc3TqnBqf1ufUnOWLOOw/Vw27b6GvKJm5BFgCcPhOU7LIc2toG1ddPOJ2w7G66bAF97FyDbCgsbxyTanlSJJqYqgBzVadCMj3Age0LE9rrquV0f6hJT3F3xdJqrZ+H/3uREC+Gjg3IpYCLwF+k5lPa68o20Q+Ai5K0TWONnosdYPhWGPlc0ZS0UYNyBHxDeAEYI+IWA98ENgBIDMvB5YBJwOrgceAN3ar2KFUIfhKMuSoe3q5/3iwgVpbe5El9Z5RX1Uy84xRtidQ25TqH31peP5+SJKqqDpvu9Ux8079RNklqAYMx5LaMdYZNFS8iczUUZaJ1mxALthwLSGL9jmNVezX1jEW7XMawIhTuUllMBRLxeu1gDnclHP9yDbP/tXOnfR6zpzli3xSSpKkvlW3rNNr/6+OIEuaMEeOperqtRFoqRcYkCdgoNWhCNveWS2Y2HG88Yc6rcjfA0ntm7N8EZcf2zt/5g3iqpJKtlgUpW4fb0iSJMkR5L7RayMF6n+OHEuS+pWJahyq2G9pa4U6pYrPf0lSfQ10A4xlINEWi4IZLiRJknqbAXmMDLiSJEm9oxvXi9liUZBOzUIxHrZXSJIktc+AXGHOsCGpzpw2TFK3GJD7nKPH6hRnrZCk4sxf3Iho59xcciE1ZUBuk73HkqQqGBhZn7O83Dp6zcCnrtedcFnJlagKvEivDY6cSZIk1YcBWZIkqU91+3qlbtx1uBeusTIgSxqVLUaSpDqxB1nSsAzGUncMXIAl9aqBXvZzbr605ErKqcXf0IIMvBiuXFNyIZIkFaRbU/FV+YK7OcsXVbLuujEgj8DRM0mSpPoxIEuSpDHp5k1avAGMekFbATki5gKXApOAL2TmxYO2Hwh8Bdi1uc8Fmbmss6UWy9FjSZKk4vTC7BUDRp3FIiImAZcB84BZwBkRMWvQbh8ArszMo4AFwGc7XagkSZL6XxEXsU7f9PURt7dTwTHA6sy8ByAilgKnAHe17JPALs3Hzwa6Nvw6f/Fkrvzolm4d3pFjSephVb44SxqKz+liXbLr423t105A3h9Y17K8HnjJoH0+BHwvIv4C2Bl41VAHioiFwEKAfSfb/ixJGp+69anW7f+3TgZGS8+5uZzzG9CH1qkbhZwBfDkzpwEnA/8QEU87dmYuyczZmTl7t0kGZEmSJLVnom8UN666ePSdmtoJyPcCB7QsT2uua/Vm4EqAzLwZ2AnYo+0qNKLh2j5sB5EkSeq8dgLybcDMiJgRETvSuAjv6kH7rAVOBIiIw2gE5Ic6WagkSZK6z5aeNnqQM3NLRJwLfJfGFG5fzMw7I+IiYEVmXg28E7giIt5O44K9N2RmdrPwbli0z2mlnn9gRHjV0v1KrUOSpLHw1tnqN209o5tzGi8btO7Clsd3Acd3tjRNxEDT/SWnfqLkSiRJZTC0dtbAqOo5N19aciUqQqcu0pMkSVIfG6n1opdu8tEJBuQe1W9PNEmSpG4bCPHzF0+e0KcoBmRJkqQ+ZavN+BiQJUmSamLRPqeVPilBFVTybcXAu6FO3XLa+YQlSZLGZtE+p7GK7s68NWf5Ii4/dmJxdd44JixwBFmSJElqYUCWJEk9Y+Wataxcs7bsMlRzBuQ+YZuIJEkayuXHntf1u+P128WAtQvI/fYDlCRJqrORst143xjULiBLkiSpHjauunhc31f74VRbEyRJ6j3jDTb9YqAP+7KS64B6fvpev/9jSZLUs6Zv+nrZJUi2WFTdYQvucxRckiT1jPH0/Y5nlLqbFx8akCVJknrUnOWLyi6hlmyxkCRJKkkv9RrrKbUOyN6LXJIkqf9M9CLPWgbkqvTsVqVOSSpKHa+mV/Vcfux5nHPzpWWXUbhu34ykHRtXXcwluz4+4eP4SiNJkqTSPTUweGCpdUANL9Lz/u6SJKkq/NRkdAOjxp0YOR5Qu4AsSZIkjaSttyURMRe4FJgEfCEzn9b5HBHzgQ8BCfw4M8/sYJ0TZj+vJEnqBKde6w3dvNviqCPIETGJxuwj84BZwBkRMWvQPjOBxcDxmfl84PzOlypJkqQBvXBR3FhUqd52RpCPAVZn5j0AEbEUOAW4q2Wfs4HLMvPXAJn5YKcLlSRJUjUMhOFOz+bRyT7jkbQTkPcH1rUsrwdeMmifQwEi4iYabRgfysxrBx8oIhYCCwH2nWzTeTdtaynZVG4dkiRJndDNlorBOpVSJwMzgROAacANEXFEZj7SulNmLgGWABy+05Ts0LklSepZVfpYWdU20Bt93QnF3Zev27NszDv1E7Cqq6cYUjv/V/cCB7QsT2uua7UeuDUzNwNrIuJnNALzbR2pUpIk9b2Va9Yyvewi1HFVfJPYTkC+DZgZETNoBOMFwOAZKq4CzgC+FBF70Gi5uKeDdQ5r/uLJXPnRLUWcSpIkddH0TV8vu4TKGjyzRhmhdOBeE0fMmPiNPorqNR7OqAE5M7dExLnAd2n0F38xM++MiIuAFZl5dXPb70fEXcBW4N2ZuaGbhY+FU7xJkiT1tm1vkEpoqRisrcaRzFwGLBu07sKWxwm8o/klSZKkDpuzfFGh/cV11tdTSThyLKlXbN5xV9Yf/V42PfsgIMouZwjJTr+5h2m3f4wdnnik7GKk2rn82PM6PiVaVVyy6+NQ4AwV7ejrgNwrvI+6pPVHv5epB81m+s6Tiei9gJyZbHh0N9bzXmbcsrjscqRaKncWivoG9KGMeic9SdLEbXr2Qezeo+EYICLYfefJzRFuSaq3vg3Ii/Y5rewSJKlF9Gw4HtCor7drlKQi+Nm/JElSRYzUqzxw7dV19xdZ0diUPX1bu/ouIHthnqQqmP53nX2t+vlf7jfqPtdefxPnXfgJtj65lbeccRoXnPvGjtYgSf2iLwKyofjpnGxdUqutW7ey6P0f4/vf+CzT9t2bF598Fq/+/Vcw61B7jqVeN/gmIN22cs3ajtzsY7CNPTZTxUj6tgdZkvSUf7/jpxwyfRoH/Y9p7LjjDiw45X/yz99dXnZZkgpUxVs+l8WALEk1cO/9D3HAfvtsW562717ce/+DJVYkqVucqGDi+qLFQpIkVVeVPnqvm/GOOg/8THvxJiDtcARZkmpg/332ZN19T13avv6XD7L/PnuVWJGkftUPb3gcQa4IL0SUNBEvfuHz+b9r1rFm7b3sv89eLP3n7/L1yz5SdllSTxsYPfUOc+3ppwkCDMg9ykAs9bd2pmXrpMmTJ/OZv3ov//PMRWx98kne9Kev5vnPPbjQGlRvK9esBejK7AjqvLH+vPph1LiVAVmSauLkE1/GySe+rOwypO30W7AqWtFTwA2Yd+onYBOwqpTTd11PB+T5ixvlXfnRLSVXUi399BGHJEnqDRtXXVyZO+FNVE8HZEmS1L8cPR6bgYHDIuczruvPqPIBeaBHRvV9EkuS6qOslgLVS+UDsiRJUr9auWYtq+jORb3DjUTbquk8yJIkSV3j7Z2rqdIjyE6FJkmS1Bl1uQCvHZUOyJJUWUtO6OzxFi4fdZc3veND/Mu//oC99tiNn173rc6eXxqDOl8zM3/xZM65eezf0211/pkMpa0Wi4iYGxF3R8TqiLhghP1Oj4iMiNmdK1GS1AlvmP9HXPv/fabsMqS+ctiC+0b9RLuX2yw2rrrYcDyEUQNyREwCLgPmAbOAMyJi1hD7TQXOA27tdJGSpIl7+UtfxG67PrvsMlRzhrHhzV88uZDRYmi0U/izGF47P4VjgNWZeQ9ARCwFTgHuGrTf/wI+Bry7oxVKkqTKc2aE3uDPoT3ttFjsD6xrWV7fXLdNRBwNHJCZ3xnpQBGxMCJWRMSKX2317niSJEnqPRMex4+IZwB/C7xhtH0zcwmwBODwnabkRM8tSZJ6X7c+yp+zfBHXnXBZV449koE2iCs/Wo3BPlspxq6dgHwvcEDL8rTmugFTgcOB5REBsA9wdUS8OjNXdKrQKlu5Zi1HzDiw7DIkSVKPGm9WGC6sG4onpp2AfBswMyJm0AjGC4AzBzZm5m+APQaWI2I58C7DcbHsKZIqpo1p2TrtjLctZvnN/8HDv3qEaS+ay4ffdQ5vPuPUwutQd61csxbAgZkaMgt0zqgBOTO3RMS5wHeBScAXM/POiLgIWJGZV3e7SEnSxH3jsx8tuwTViCOYo5u/eDIr13TmWPNO/URnDiSgzR7kzFwGLBu07sJh9j1h4mVJkiRpNL4R6Y5K3klv4OOjKqlizZIkjYVhrRj+O3dfJQOynmK/kSRJUmcZkCVJkirAkePiVCYgD9znvFPN7JIkqTMMbt3lv2/x2rmTniRJUt+7/NjzuPzY88ouQz2gEiPIA6PH2p79x1J1HfH913X0eCtP+uqI29fdez+vO+9CHnh4AxHBwj97Dee95cwRv0dS+fxbX45KBGRJ0sRMnjyJv/ng2zn6iMPY+NtHedHcP+Okl7+UWYceVHZpUmU5gNe/DMiSVAP77r0n++69JwBTn7Uzh82cwb33P2hAlnqUI8flMiBLUs38fN193PHTu3nJUYeXXYoqzovH1K8MyJJUI7999DFOP/tdfOrD72SXqc8quxxVkKG4+y7Z9fGyS6g9Z7GQpJrYvHkzp5/9Lv7stJN5zcknll2OKmbjqot7PhzPWb6o7BLUJxxBlqQayEze/M6LOOyQGbzjrWeVXY6kIfT6G5A6MSBLUglGm5at02667Uf8w///HY447BBeeNICAD5ywbmcfOLLCq1DqgvnU642A3IFeWWrpLF62TFHkffeXnYZklQJ9iBLkqQR+dG/6sYR5Ipx9FiSVARD8fBWrlnb8WP69723GJArwl8cSZL6j3/fe5MtFpIkSS3Ge4Hd5cee58V5fcKAXAG+u5QkqTcYgOvBFoseZSiWJBXNvuPRdaP/WL3HgCxJJVg158yOHu+w60Z+U71p0+94+elv4Xe/e4ItW7fyx39wIh9+1593tAZVk6FYerq2AnJEzAUuBSYBX8jMiwdtfwfwFmAL8BDwpsz8RYdrlSSN0zOfuSPXXfl5nrXz77F582ZedtqbmffK43npi44su7RaWrlmLUfMOLC08/dTKB64vfR1J1xWciXtu2TXxxsP+ujn0G9G7UGOiEnAZcA8YBZwRkTMGrTbHcDszDwS+Efg450utE7mnfqJskuQ1Gcigmft/HsAbN6yhc2btxARJVelMvRTOO6k+YuL+VDdf/9qaOfZcAywOjPvAYiIpcApwF0DO2Tm9S373wKc1ckiJUkTt3XrVl40989Y/fN1LHrDfF5y9BFll1Q79q/2nnYvuvNnVy/tzGKxP7CuZXl9c91w3gxcM9SGiFgYESsiYsWvtm5pv0pJ0oRNmjSJH31/KetXXMu/33EnP/3P1WWXVCsGLKk6Ovp5QkScBcwGXjHU9sxcAiwBOHynKdnJc0uS2rPrs6fyyuNnc+3yH3L48w4puxwVxI/2pfa1M4J8L3BAy/K05rrtRMSrgPcDr87M33WmPElSJzy04dc88puNADz++Ca+f8MtPO/g6eUWJUk9qp0R5NuAmRExg0YwXgBsNz9RRBwFfB6Ym5kPdqo4P46S1K9Gm5at0375wEO8/vwPsvXJrTz5ZDL/j07iD096eaE1SL3Gm35oOKMG5MzcEhHnAt+lMc3bFzPzzoi4CFiRmVcDlwDPAr7VvCp6bWa+uot1S5LG4MhZh3LH975RdhmSVAlt9SBn5jJg2aB1F7Y8flWH66ol754nSeok+47LN/Az2Db3sSrBO+lJkiR1mG9Oqq2di/QkSZJqa9E+p5VdggpmQO4RtldIkjrJEUxp/GyxkCSpT2xcdTHTyy6ij4x3Ni3fnFSfAVmSpIozkEmdZUCWpBJcdtGvO3q8RRc+p639tm7dyux5Z7H/PnvyL1/9u47WoHIYjsfO+Y81GgNyyead+omyS5BUI5d+4RscNnMG/73xt2WXIkk9y4v0JKkm1t/3AN/5Pz/gLWecWnYpktTTHEGWpJo4/4Of4OMfOI+Nv32s7FLUAbZWjE+77RVO7VZvBmRJqoF/+f4N7LXHbrzoyFks/+GKssvROBmKe5s/n/5hQJakGrhpxY+5+nv/xrLrbmTT757gvzc+yll/8X6+9um/Lrs0qRDzF3c38ngr6f5iD7Ik1cBHF/8F6//jWn5+63dY+tmPMuf42YZjSRqGI8iSVIJ2p2WTBvjxvVQcA7Ik1cwJx83mhONml12GRmEglspjQC6Bcx9LkoZjMO48bwyisTIgS5KkvtTtC/PUv7xIT5IKkWRm2UWMqFFf79Y4fdPXyy6h6xw97hxHjTURvrUqWB1e4CU93U6/uYcNj+7G7jtPJiLKLudpMpMNj25hp9/cU3YptWQwlnqLAVmSCjDt9o+xnvfy0LMPAnovIEOy02/uYdrtHyu7EGlcLj/2PC4/djLgyLEmzoAsSQXY4YlHmHHL4rLLqKSqf/K2cs3askuQNEb2IEuSVCLbK6Te4wiyJEmqrG7MVOGovwzIBan6R4SSVIaqv3aOFLQcOZZ6V5Q17VBEPAQ8CjwM7NFc3fp48PJEHnusatXisapbi8fyeeCxql2Lx6puLR5rfI//MzPnMkhpPciZuSfwcGbOplHkdo9H2jbWxx6rWrV4rOrW4rF8Hnisatfisapbi8ca3+McIhyDF+lJkiRJ2zEgS5IkSS0ml3z+JYP+O/jxSNvG+thjVasWj1XdWjxWecfqpVo8VnVr8VjVrcVjTexY25R2kZ4kSZLUi2yxkCRJkloYkCVJkqQWBmRJkiSphQFZkiRJalFIQI6IT0XE/IjYv4jzSZL6X0R8s+waJPWnoqZ5Ww2cCnw8IgB+2Py6CfhxZj5ZUB19LSI+DbROS5I07hhzfWbeWE5VqpKIeC6wEHhec9Uq4IrMvLu8qqRhHVt2ASpGRJwEvCczTyq7FtVD4dO8RcR+wHHNr1cDe2XmLoUW0aci4vVDrN4NmA98MzM/VWxFqpKIOBb4J+DzwB1AAEcBZwOvycxbSixPepqIWJuZB5ZdhzonIuYAlwP7AVcBHwO+ROP16K8z85/Kq051UlhAjsbQ8RE0gvHxwCzgIeDmzPxwIUXUVERMAX6YmUeVXYt6V0RcA3wsM5cPWv8K4ILMnFdKYaq1iDh6uE3Av2TmvkXWo+6KiDuAtwM3A/OAr9F4/flMqYWpdgoJyBHxfWAX4EfALcAtmbmq6yfWNhFxhwFZI4mIn2XmocNsuzszn1t0TVJEXD/S9sx8ZVG1qPsi4vbMPLpl2dcelaKoHuR7gCOBmcAG4OGIeCgzHy7o/LUVEZOB1wLry65FPW/jCNseLawKqYUBuHZ2jYjXtCxPbl22xUJFKbQHOSJ2AV5Ko83ipcCewE8zc6jeWY1RRGxk+4v0AB4H/g04PzPvK74qVUVEPAgsHWoTMD8z9y64JImIeE9mfrz5+E8y81st2z6Sme8rrzp1WkR8aYTNmZlvKqwY1VrRAfmZwItp9CAPhOQHM/OIwoqQNKRhLvLcJjO/UlQt0oDWj9yH+Ph9u2VJ6pRCWiwi4pM0AvFMGlfH30zjKtXXZ+YjRdRQBxFxVmZ+rfn4+My8qWXbuV7koFH8mMa0i8VObSONLIZ5PNSyKi4i3jFo1cB0pTdm5poSSlJNFXUnvTXA24A9M/PEzPxAZl5jOO641heWTw/a5sdSGs0XgA0R8f2I+HBE/H5ETC27KNXe4Lndh9um/jB10NcuwGzgmohYUGZhqpeiZrH4H8Ajmfmb5vIradw45BfAZzLzia4XUQOtM1UMnrXCWSzUjoj4PeAYnpqr/MXA/cBNmfm2MmtTPUXEVhoXiQYwBXhsYBOwU2buUFZtKk5E7Ab8qy01KkpRI8hXAjsDRMQLgW8Ba4EXAJ8tqIY6cKRFE5KZjzXnQb4U+CRwGY3f3bll1qX6ysxJmblLZk7NzMnNxwPLhuOayMxfYUuNClTUNG9TWmZQOAv4Ymb+TUQ8g8bcyOqM50XET2i8iBzcfExz+aDyylIVRMSZNEaNXwj8DrgNuBV4WWbeX2JpqrGI2Ak4BzgE+AmNvx9byq1KRWt+8vzrsutQfRQVkFvf9c0BFgNk5pPNO+ypM14A7A2sG7T+ABofk0sj+TxwN40LaG/IzJ+VXI8E8BVgM/AD4GTg+cB5pVakromIlTz9E8/dgPuA1xVfkeqqqIB8XURcCfwSeA5wHUBE7AtsKqiGOvgksDgzf9G6sjn/9CeBPyqlKlXFrjTeZB0HfCginkvjd/ZmGreEv67E2lRfswamAo2Ivwf+veR61F1/OGg5gQ2Z6c2KVKiiAvL5wJ8C+9L4uHZzc/0hNN4ZqjP2zsyVg1dm5sqImF5CPaqQzNwK3N78+kxE7A38CY3f34uASeVVpxob+HtBZm7xQ8e+9wBPtdSsBP7elhqVoZCA3JxXdSlARBwVEefT+MO7BvhUETXUxK4jbJtSVBGqpog4kqdmrzgO2JHG6PGngZtG+Fapm14QEf/dfBzAlOZy0Pjzskt5pakLWltq5gGzsKVGJSjqRiGHAmc0vx4GvkljirlXFnH+GlkREWdn5hWtKyPiLcB/lFSTquPLwI3ANcAHMnNtueVIjVksyq5BhbKlRj2hqHmQn6TxbvDNmbm6ue6ezHRmhQ5qfiT+beAJngrEs2mMBJ7mTAQaSUScCkwbuONiRNwK7Nnc/J7M/MeyapNUD95OXL2iqIB8KrAAOB64lka7xRcyc0bXT15DzelwDm8u3unFVWpHRNwELMjMdc3lHwEn0pgH+UuZeWKJ5UmqgZYbw8D2N4expUaFKqoH+SrgqojYGTiFxkU/e0XE54BvZ+b3iqijLjLzeuD6sutQ5ew4EI6bbszMDTRuP71zWUVJqg9batQrChlBHvLEEc+hcaHenzoyJZUvIlZn5iHDbPuvzDy46JokSSpDUbeafprM/HVmLjEcSz3j1og4e/DKiHgrXigjSaqR0kaQJfWWiNgLuIrGbaZvb65+EfBM4NTMfKCk0iRJKpQBWdJ2ImIOjdv5ghd5SpJqyIAsSZIktSitB1mSJEnqRQZkSZIkqYUBWZIkSWphQJYkSZJa/D9hZq28vBXsogAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\" Check ADMIXTURE results \"\"\"\n",
    "%pylab inline\n",
    "\n",
    "# The code below makes ADMIXTURE bar plots based on your results\n",
    "# It will fail if your ADMIXTURE run was not complete or if you did not\n",
    "# output results to the specified output files!\n",
    "# Please ensure you get plots (and the notebook validates)\n",
    "# to get full points.\n",
    "\n",
    "sampleinfo = pd.read_csv(\"~/public/1000Genomes/igsr_samples.tsv\", sep=\"\\t\")\n",
    "sample_to_pop = dict(zip(list(sampleinfo[\"Sample name\"]), list(sampleinfo[\"Population code\"])))\n",
    "samples = [line.split()[0] for line in open(\"ps2_admixture.pruned.fam\", \"r\").readlines()]\n",
    "pops = [sample_to_pop.get(item, \"NA\") for item in samples]\n",
    "\n",
    "fig = plt.figure()\n",
    "fig.set_size_inches((10, 8))\n",
    "\n",
    "plotind = 1\n",
    "for K in [3, 5]:\n",
    "    ax = fig.add_subplot(3, 1, plotind)\n",
    "    \n",
    "    data = pd.read_csv(\"ps2_admixture.pruned.%s.Q\"%K, delim_whitespace=True, header=None)\n",
    "    cols = list(data.columns)\n",
    "    data[\"sample\"] = samples\n",
    "    data[\"pop\"] = pops\n",
    "    data = data.sort_values([\"pop\"]+cols)\n",
    "    data.plot.bar(stacked=True, ax=ax, width=1)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom')\n",
    "    \n",
    "    # Only plot label for first sample in each pop\n",
    "    xticklabels = []\n",
    "    currpop = \"\"\n",
    "    for i in range(data.shape[0]):\n",
    "        if data[\"pop\"].values[i] == currpop:\n",
    "            xticklabels.append(\"\")\n",
    "        else:\n",
    "            xticklabels.append(data[\"pop\"].values[i])\n",
    "            currpop = data[\"pop\"].values[i]\n",
    "    ax.set_xticklabels(xticklabels)\n",
    "    \n",
    "    plotind += 1\n",
    "fig.tight_layout()\n",
    "\n",
    "\n",
    "print(\"These are the actual results from the admixture tool \" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "57c7849b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starting reading \n",
      "Genotype matrix shape: (26499, 462)\n",
      "(26499, 462) [[0.0000000e+00 4.2215400e+05]\n",
      " [1.0000000e+00 1.9223000e+04]\n",
      " [2.0000000e+00 1.0037297e+07]\n",
      " [          nan 1.7638640e+06]]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "### this is the code for reading from the .bed file \n",
    "print(\"starting reading \")\n",
    "\n",
    "def read_bed_file(bed_file_path, num_individuals, num_snps):\n",
    "    \"\"\"Read genotype data from a .bed file.\"\"\"\n",
    "    # Open the .bed file in binary mode\n",
    "    with open(bed_file_path, 'rb') as bed_file:\n",
    "        # Skip the header\n",
    "        bed_file.read(3)\n",
    "        \n",
    "        # Calculate the number of bytes to read per row\n",
    "        bytes_per_row = (num_individuals + 3) // 4\n",
    "        \n",
    "        # Initialize an empty array to store the genotype data\n",
    "        genotypes = np.zeros((num_snps, num_individuals), dtype=float)\n",
    "        \n",
    "        for snp_index in range(num_snps):\n",
    "            # Read a row of data\n",
    "            row_data = bed_file.read(bytes_per_row)\n",
    "            byte_array = np.frombuffer(row_data, dtype=np.uint8)\n",
    "            \n",
    "            # Decode the genotypes from the byte array\n",
    "            for i in range(num_individuals):\n",
    "                byte_index = i // 4\n",
    "                if byte_index >= len(byte_array):\n",
    "                    break\n",
    "                \n",
    "                # Extracting bits for one individual\n",
    "                # Bits are stored in reverse order for each individual\n",
    "                bit_index = (i % 4) * 2\n",
    "                genotype_code = (byte_array[byte_index] >> bit_index) & 0b11\n",
    "                \n",
    "                # Convert the binary genotype code to the actual genotype\n",
    "                if genotype_code == 0b00:\n",
    "                    genotypes[snp_index, i] = 0  # Homozygous first allele\n",
    "                elif genotype_code == 0b01:\n",
    "                    genotypes[snp_index, i] = 1  # Heterozygous\n",
    "                elif genotype_code == 0b11:\n",
    "                    genotypes[snp_index, i] = 2  # Homozygous second allele\n",
    "                elif genotype_code == 0b10:\n",
    "                    genotypes[snp_index, i] = np.nan  # Missing genotype\n",
    "                else:\n",
    "                    raise ValueError(f'Unexpected genotype code: {genotype_code}')\n",
    "                    \n",
    "    return genotypes\n",
    "\n",
    "# Parameters: Update these based on your .bed file\n",
    "bed_file_path = \"ps2_admixture.pruned.bed\"\n",
    "num_individuals = 462  # Update with the correct number of individuals in your dataset\n",
    "num_snps = 26499  # Update with the correct number of SNPs\n",
    "\n",
    "# Read the genotype data from the .bed file\n",
    "genotype_matrix = read_bed_file(bed_file_path, num_individuals, num_snps)\n",
    "print('Genotype matrix shape:', genotype_matrix.shape)\n",
    "\n",
    "(unique, counts) = np.unique(genotype_matrix, return_counts=True)\n",
    "\n",
    "frequencies = np.asarray((unique, counts)).T\n",
    "\n",
    "# Change binary genotype to ped+map format\n",
    "#plink --bfile ps2_admixture.pruned.bed --cow --nonfounders --allow-no-sex --recode --out ps2_admix_trial\n",
    "dimensions = genotype_matrix.shape\n",
    "print(dimensions, frequencies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1fabda84",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starting now \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:   0%|          | 1/1000 [00:01<24:07,  1.45s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5645                  Q L1 Loss = 0.4081\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:   5%|▌         | 51/1000 [01:15<23:53,  1.51s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2453\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  10%|█         | 101/1000 [02:26<20:46,  1.39s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2516\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  15%|█▌        | 151/1000 [03:40<21:13,  1.50s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2495\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  20%|██        | 201/1000 [04:53<19:40,  1.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2495\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  25%|██▌       | 251/1000 [06:08<18:49,  1.51s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2403\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  30%|███       | 301/1000 [07:22<16:59,  1.46s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2501\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  35%|███▌      | 351/1000 [08:36<15:39,  1.45s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2505\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  40%|████      | 401/1000 [09:49<14:44,  1.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2427\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  45%|████▌     | 451/1000 [11:02<13:06,  1.43s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2413\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  50%|█████     | 501/1000 [12:16<12:07,  1.46s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.239\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  55%|█████▌    | 551/1000 [13:34<11:37,  1.55s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2465\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  60%|██████    | 601/1000 [14:53<10:30,  1.58s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2419\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  65%|██████▌   | 651/1000 [16:11<09:27,  1.63s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2424\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  70%|███████   | 701/1000 [17:32<08:27,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2448\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  75%|███████▌  | 751/1000 [18:57<06:50,  1.65s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2437\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  80%|████████  | 801/1000 [20:21<05:46,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2507\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  85%|████████▌ | 851/1000 [21:49<04:24,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2433\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  90%|█████████ | 901/1000 [23:17<02:58,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2483\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations:  95%|█████████▌| 951/1000 [24:48<01:28,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P L1 loss = 0.5624                  Q L1 Loss = 0.2419\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "EM Iterations: 100%|██████████| 1000/1000 [26:16<00:00,  1.58s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "#### run this code \n",
    "print('starting now ')\n",
    "targ_Q = np.array(pd.read_csv(\"ps2_admixture.pruned.3.Q\", delim_whitespace=True, header=None))\n",
    "targ_P = np.array(pd.read_csv(\"ps2_admixture.pruned.3.P\", delim_whitespace=True, header=None))\n",
    "def naive_admixture(genotypes, num_individuals, num_snps, K=3, num_iterations=1000):\n",
    "    # Initialize P (allele frequencies) randomly for each SNP for each population\n",
    "    P = (np.random.rand(num_snps, K)/5)+0.4\n",
    "    \n",
    "    # Initialize Q (ancestry proportions) randomly for each individual\n",
    "    Q = (np.random.rand(num_individuals, K)/5)+0.4\n",
    "    \n",
    "    # Normalize P and Q so they sum to 1 across the K populations\n",
    "    P /= P.sum(axis=1, keepdims=True)\n",
    "    Q /= Q.sum(axis=1, keepdims=True)\n",
    "    N_limit = 90  # EM algorithm with progress bar from tqdm\n",
    "    for iter in tqdm(range(num_iterations), desc=\"EM Iterations\"):\n",
    "        # E-step: Estimate ancestry fractions\n",
    "        for i in range(num_individuals):\n",
    "            stoch_indices = np.random.choice(num_snps, N_limit, replace = False)\n",
    "            curr_genotypes = genotypes[stoch_indices,i]\n",
    "            probs = P[stoch_indices,:] ** np.tile(curr_genotypes.reshape(-1,1), (1,K)) * (1-P[stoch_indices,:]) ** (2-np.tile(curr_genotypes.reshape(-1,1), (1,K)))\n",
    "            good_indices = np.logical_not(np.any(np.isnan(probs), 1))\n",
    "            probs = probs[good_indices]\n",
    "            good_indices = np.logical_not(np.any(probs==0, 1))\n",
    "            probs = probs[good_indices]\n",
    "            probs = probs / probs.sum(1).reshape(-1,1)\n",
    "            \n",
    "            log_probs = np.sum(np.log(probs), 0)\n",
    "            log_probs = log_probs - log_probs.max()\n",
    "            Q[i] = np.exp(log_probs) + 1e-10\n",
    "            Q[i] = Q[i] / Q[i].sum()\n",
    "#             print(Q[i])\n",
    "#             asdf\n",
    "#             print(curr_genotypes)\n",
    "#             print(curr_genotypes.shape)\n",
    "#             print(P.shape)\n",
    "#             asdf\n",
    "            \n",
    "            \n",
    "            \n",
    "#             for j in range(num_snps-1):\n",
    "#                 genotype = genotypes[j, i]\n",
    "#                 probs = P[j] ** genotype * (1 - P[j]) ** (2 - genotype)\n",
    "#                 if np.any(np.isnan(probs)):\n",
    "#                     continue\n",
    "#                 Q[i] *= probs\n",
    "                \n",
    "            # Normalize Q for this individual\n",
    "        \n",
    "            \n",
    "\n",
    "        # M-step: Update allele frequencies\n",
    "#         for j in tqdm(range(num_snps), desc=\"M-step\", leave=False):\n",
    "        for j in range(num_snps):\n",
    "            for k in range(K):\n",
    "                curr_genotypes = genotypes[j]\n",
    "                good_indices = np.logical_not(np.isnan(curr_genotypes))\n",
    "                curr_Q = Q[:,k]\n",
    "                curr_Q = curr_Q[good_indices]\n",
    "                curr_genotypes = curr_genotypes[good_indices]\n",
    "                P[j, k] = np.sum(curr_Q * curr_genotypes) / (1e-10 + (2 * np.sum(curr_Q)))\n",
    "            # Normalize P for this SNP\n",
    "            P[j] /= P[j].sum()\n",
    "        \n",
    "        if iter % 50 == 0:\n",
    "            l1_P = np.abs((targ_P - P)).mean().round(4)\n",
    "            l1_Q = np.abs((targ_Q - Q)).mean().round(4)\n",
    "            print('P L1 loss = {}                  Q L1 Loss = {}'.format(l1_P, l1_Q))\n",
    "\n",
    "    # Save results to files\n",
    "    np.savetxt('result.Q', Q, fmt='%f')\n",
    "    np.savetxt('result.P', P, fmt='%f')\n",
    "    \n",
    "    return Q, P\n",
    "\n",
    "# Assuming 'genotype_matrix' is already defined and loaded with your actual genotype data\n",
    "# Running the naive admixture analysis\n",
    "### These L1 losses are not very accurate as the original population codes are getting changed. \n",
    "Q, P = naive_admixture(genotype_matrix, num_individuals, num_snps)\n",
    "print('done')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cc1dbd2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n",
      "These are the actual results from the admixture tool \n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAADKCAYAAAC4/LZUAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAAsTAAALEwEAmpwYAAAYuUlEQVR4nO3de9RddX3n8feXBCSDAcodCU4SCRSUKncvrHpBppE1BcWWm9QbkHbajChigbGLQTrjiDOOOgU7RqW1dlm8TItZFaV2AbVSoEnBysXSSRPMhSIQgVEslYTv/HHOAycPz+U8z7PPvr5faz2Ls/fZzz7fJ/z2b3/O7/z2PpGZSJIkSerZqeoCJEmSpDoxIEuSJEkDDMiSJEnSAAOyJEmSNMCALEmSJA2YX9ULLzxyYS6+eHFVL996d2/YCMCRS15ccSWq0lg7mC3bj6SizaRfsg/STM3ovHfFEwAx0VOVBeRtP9lW1UtLhZtrEJWkUbBvUp00qT1WFpA1Ok1qgHXgv5ekNrFPU1Xa1PYMyC3SpoZZJP9dJDWd/ZjqqM3t0oCsVmvzwSupeeyT1DRdbbMG5JboagPu6t8tqRnso1Q3RbfJp3fZk81HX8JTeyxlkuvdaiDZ9Yn1LLrzKnb+2eND/YYBWY3iyUbqrrs3bKztXQ3sm1RXo26bm4++hIVLj2XxbvOJqGdAzky2PrkXm7mEJbdfNtTvGJDVGJ6AJNWRfZPqpsw2+dQeS2sdjgEigr13m88jeywd+ncMyJIkzYLBWFWrRxuMWofjMb0ah6/TgKzaq0cHIKkqVfcBVb++NMj2WA4DcoN14SDpwt8oSdIwmnBOXPy/Hix0fw+850VDbffNm2/lwsv/B9uf2c75Z7+FS1e+a06va0CWJEmqsSYE4ypt376d3/rgVXzrTz7FogP357hTzuXUf/dajjh0+DnH4xmQVTt2BJLq0g/UpQ51l21wen971z0csngRS//tIgDOOu2X+NqNtxiQ1Q52ApKgHn1BHWpQt9kGh7floUc4+EUHPLu86MD9uOOue+a0TwOyKuPBL6mO7JtUFdtefRiQG8aDR5JGxz5WZbK9FeOgA/Zl04MPPbu8+Z8f5qAD9pvTPg3IDeFBJEmjYx8rNddxr3gp/3fDJjZs3MJBB+zHdV+7kS9e8+E57dOALEmqBUOquqLNbX3Y27IVaf78+Vz9Xy7hl875LbY/8wzvPvNUXnrYS+a2z4Jq0wi1+UCSpKrZx6pMtrfROOWkEznlpBML299Ohe1JkiRJagFHkFUJ30FLkrrE816zGJA1cnYKkqQu8zzYPE6xkCRJkgYYkDVSvmuWJElNY0CWJEmSBjgHuaaaPvLa9PolSdIsrHpdsftbccu0m7z7oiv487/8a/bbZy/uuekrhbysAbmGmhoum1q3pG6yz9Ko2LbK9c4zfpmV7zqTt194eWH7dIqFJKlyBgpJs/WLrzyGvfbco9B9DhWQI2J5RNwfEesi4tIJnn9xRNwcEXdFxPci4pRCq5QkqUAGcklTmXaKRUTMA64BTgY2A2siYnVm3jew2e8AX87M34+II4AbgMUjqLfVmthhN7FmSfVhHyKpjoaZg3w8sC4z1wNExHXAacBgQE5g9/7jPYAHiyxS9eHJTJKk5/P82C7DBOSDgE0Dy5uBE8ZtcwXwFxHxH4HdgDcWUp1qxYNfkqQdeW5sp6LuYnE28IeZ+bGIeBXwhYh4WWY+M7hRRKwAVgDsvPfOBb20JEmSamGI27IV7ezfvIxbbvs7Hv3R4yw6Zjkfuvg3OO/sN89pn8ME5C3AwQPLi/rrBp0HLAfIzNsiYldgH+DhwY0ycxWwCmDBkgU5y5pbyXegkiQ1i+fueviTT/23wvc5zF0s1gDLImJJROwCnAWsHrfNRuAkgIg4HNgVeKTIQtvq7g0bPcAkSZJqZNoR5MzcFhErgRuBecC1mXlvRFwJrM3M1cD7gc9ExPvoXbD3zsx0hLglDPCSJD3H82L7DTUHOTNvoHfrtsF1lw88vg94TbGlqQ7sBCRJXee5sHv8qukKeKBJUjXsfyUNw4AsSZI0Ad9QdddQXzUtSZIkdYUjyCXz3agkSfXn+Xp2jvzW2wvd390n/9G022za8hBvv/ByfvjoViKCFW87nQvPP2dOr2tAliRJUmPNnz+Pj/3n93H0kYfz4588yTHL38bJv/hKjjh06ez3WWB9moLvRCWpWvbDGpZtpVkO3H9fDtx/XwAWvnA3Dl+2hC0PPWxAliRpMoYdqTse2PQgd91zPycc9bI57ceL9CRJlTC4SirST578KW+94GI+8aH3s/vCF85pX44gl8CTgCRJ9ef5urmefvpp3nrBxbztLadw+iknzXl/jiBLklrLwKNh2VaaKzM57/1XcvghS7jo188tZJ+OIEuSJKkQw9yWrWi3rvkuX/g/X+fIww/hFSefBcCHL13JKSedOOt9GpA1Id9JSxol+xhJRTnx+KPILXcWuk8D8gh5ApCkHdkvSmoC5yDreTyBSZKkLjMgS5IkaZaSzKy6iGn1ahy+TqdY6FmOHEtqC/szDcN2Mne7PrGerU/uxd67zSciqi5nQpnJ1ie3sesT64f+HQOyJKlVDD1SeRbdeRWbuYRH9lgK1DMgQ7LrE+tZdOdVQ/+GAVmS1AoGY6l8O//scZbcflnVZRTOOciSJEnSAAPyiDiSIUnlsc+VVCSnWEiSSmGIVV3YFjUdA7IkaaQMI5KaxoBcIE8CklQ++15JRRtqDnJELI+I+yNiXURcOsk2Z0TEfRFxb0R8sdgyJUmSpHJMO4IcEfOAa4CTgc3AmohYnZn3DWyzDLgMeE1mPhYR+42qYElSc4xydNeRY0mjMswUi+OBdZm5HiAirgNOA+4b2OYC4JrMfAwgMx8uulCNlicaSUWyT5HUZMNMsTgI2DSwvLm/btChwKERcWtE3B4RyyfaUUSsiIi1EbF2+4+3z65iSVLnGcAljVJRF+nNB5YBrwMWAd+OiCMz8/HBjTJzFbAKYMGSBVnQa0uSasDQKqkthhlB3gIcPLC8qL9u0GZgdWY+nZkbgH+kF5glSZJqwzdyGsYwAXkNsCwilkTELsBZwOpx21xPb/SYiNiH3pSL9cWVKUlSL9wYcCSN2rQBOTO3ASuBG4HvA1/OzHsj4sqIOLW/2Y3A1oi4D7gZ+EBmbh1V0ZIkSdKoDDUHOTNvAG4Yt+7ygccJXNT/6SRHNCRJktrBb9KTJEmt50CWZmKob9KTJGkqhg/Vme1TM2VAliRJkgYYkCVJkqQBzkHuOD92kjQX9iGqM9unZssRZEmSJGmAAVmSJEkaYECWJEmSBjgHeY6c3yRJUv14ftZcGJBnoQkH3eKnvgjAQi6tuBKVaez/e1FsP5Jmq4j+yD5Iw5pNe3tgiucMyC0y08bRhKBfhqJDpdQV9iH1Y3+mqrSt7RmQ1WptO2AlNZt9kprAdmpAbo2uNOau/J2Snm/xU1/kgV3PqbqMCdk3qa5sm7NjQFZjeJBLqhP7JFXFtjd6BuQG68IB0oW/UWqiMucfV90PVP360hjbYnkMyKodOwBJdekH6lKHuss2WA0DsirjQS9pIlX3DVW/vrrLtlcfBuSG8KCRVAdtvbWbfWx7NKmN2u7qy4DcAB5Akrqgqr7OPlZlsr01gwF5hpr0zrSO7Bik5rL/k2bP81+zGJCH5Ilh9uwUpGaz/5PmxvNg8xiQh9C2k0OXbs8kSZI0U0MF5IhYDnwSmAd8NjM/Msl2bwW+ChyXmWsLq7JjRhkqRxmODcNS+7RtgGCQfVb71KW92raab6fpNoiIecA1wJuAI4CzI+KICbZbCFwI3FF0kZKkdis7UBhgJE1lmBHk44F1mbkeICKuA04D7hu33e8CVwEfKLTCDmnCyLEnFUlFsk9RW9iW22WYgHwQsGlgeTNwwuAGEXE0cHBmfj0iJg3IEbECWAGw8947z7zaktXlo5q68OCXJGlHnhvbac4X6UXETsD/BN453baZuQpYBbBgyYKc62u3gQeWpDpygECanufw9homIG8BDh5YXtRfN2Yh8DLglogAOABYHRGneqFe83nwS91jOJZ25Lmwe4YJyGuAZRGxhF4wPgs4Z+zJzHwC2GdsOSJuAS42HE/OA02SqmH/q5mwvXTXtAE5M7dFxErgRnq3ebs2M++NiCuBtZm5etRFtokHm6S6cuRYeo7n624bag5yZt4A3DBu3eWTbPu6uZfVPh5oklQt+2ENy7Yiv0lPklTJ6HFZIcSwI2mmDMiTKPJkYecsqa7aPq3C/lfDsq1okAFZE7KjkDRK9jGS6syAPEJNPQE0tW5JM9PmaRWSNBcG5HHa/nHjVDxxSWoL+zMNw3aiyRiQO6DLoV9Sdxh2JBVlp6oLqBMvzJOkZrLPlVQkR5AlSSNleFXd2CY1HQNygTzgJKl89r2SimZAxjm64AlG6oqy+rsy+hT7LUmjYkCWJBWizMBqONZ4DnapSAZkSVJjGIw13kyDsW1Iw+hkQB7Fu0wPOEmSyuWosUalkwFZklQsBwlUZ7ZPzZT3QZYkSZIGOILccb6rltpt1B9B24eozmyfmi0DsiS1jPMyJWluDMiz4DtSSZLqzXO15sI5yB1m5yFpLuxDJLVV5wKyHz1KkiRpKp2ZYmEwltQ2VfZrjh5LarNOBGTDsaQ2MRhLk7ONqghDBeSIWA58EpgHfDYzPzLu+YuA84FtwCPAuzPzBwXXWikPOElVqNMbfPtBSV0xbUCOiHnANcDJwGZgTUSszsz7Bja7Czg2M38aEf8B+Chw5igKrkLbTgpt+3skjZZ9hqSuGWYE+XhgXWauB4iI64DTgGcDcmbePLD97cC5RRap4niik5qhLiPH9hlqEturijJMQD4I2DSwvBk4YYrtzwO+MdETEbECWAGw8947D1ni7M31BNOmA61Nf4vUdnUJx5LUVYVepBcR5wLHAq+d6PnMXAWsAliwZEEW+dqS1HR1Cca+oZbUdcME5C3AwQPLi/rrdhARbwQ+CLw2M/+1mPIkqd3qEorBYCxJY4YJyGuAZRGxhF4wPgs4Z3CDiDgK+DSwPDMfLrxKSWqROoXiMYZjSXrOtAE5M7dFxErgRnq3ebs2M++NiCuBtZm5GvjvwAuBr0QEwMbMPHWEdY+cJwtJXWBfJ0nPN9Qc5My8Abhh3LrLBx6/seC6KuPJQtIo1Wn02P5OkibWiW/SG1ZbTxZt/bukpqhTKJYkTa+1AdkTUs/dGzayuOoipA6yD5Kk5mptQJ4JR1glFaEJodj+Tm1l21aROhmQPYgkSWo+z+calZ2qLkCS2qAJo8eSpOF0cgRZkopSx2DsqJokzY0BWZJmqY7hWOqCsWNvcbVlqMU6E5AdUZE0F3UMw/Zr6qI6Hotqn04EZE8ikprK/kuSyteKgOy7SUmjUGXfYjCWpOq0IiBPxhOMpNmqIhzbZ0lSPTQ+IDt6LKlIZfQpBmFJqjfvgyxJJTIcS1L9NX4EeTKehHr8d5Cez1FiSdJUWhuQJWkiowzHhmJJagcDsqTWGvVIsYFYKo/XHKlMjQ7IHiySxhtVv2AYlqozeFx7LKoMjQzI40+AHiySoPhwbN8iSd3UuIBsOJY0yFAsSSpa4wKyJM01FBuCJUlTMSBLaoTZhGKDsNQeHs8qU2MCshP0pe6ZaSi2b5DayWNbZRsqIEfEcuCTwDzgs5n5kXHPvwD4I+AYYCtwZmY+UFSRhmOp3WYShO0DJEmjNm1Ajoh5wDXAycBmYE1ErM7M+wY2Ow94LDMPiYizgKuAM+dSmBfjSe01bCD2uJckVWGYEeTjgXWZuR4gIq4DTgMGA/JpwBX9x18Fro6IyMycSTGOFEvtMVUI9viWNIyxfmRxtWWog2K6DBsRvwIsz8zz+8u/BpyQmSsHtrmnv83m/vI/9bd5dNy+VgAr+ouH0ZuO8SiwT3/d4OPxy3N57L6aVYv7am4t7st24L6aXYv7am4t7mt2j/8hM5czzk7jV4xSZq7KzGP7PwuBRzPz2H6ROzye6rmZPnZfzarFfTW3FvdlO3Bfza7FfTW3Fvc1u8c5QTiG4QLyFuDggeVF/XUTbhMR84E96I0OS5IkSY0yTEBeAyyLiCURsQtwFrB63DargXf0H/8KcFPmzOYfS5IkSXUwf7oNMnNbRKwEbqR3m7drM/PeiLgSWJuZq4HPAV+IiHXAj+iF6GGsGvff8Y+nem6mj91Xs2pxX82txX1Vt6861eK+mluL+2puLe5rbvt61rQX6UmSJEldUupFepIkSVLdGZAlSZKkAQZkSZIkaYABWZIkSRpQSkCOiE9ExBkRcVAZrydJar+I+FLVNagcEXFyRHyr6jrUHdPe5q0g64A3Ax+NCIC/6f/cCvx9Zj5TUh2tFhG/BwzeliTpfWPMzZn5nWqqUpNExGH0vg7+5/urvg98JjPvr64qaVKvqroAFSsi3gD8b+BFwPXAVcAfAAH81+oqU9eUfpu3iHgR8Or+z6nAfpm5e6lFtFREvGOC1XsBZwBfysxPlFuRmiQiXgX8KfBp4C56J6SjgAuA0zPz9grLk54nIjZm5ourrkPFiYi7gPcBtwFvAv4YuDQzr660MHVOaQE5ekPHR9ILxq8BjgAeAW7LzA+VUkRHRcQC4G8y86iqa1F9RcQ3gKsy85Zx619L7wT1pkoKU6dFxNGTPQX8eWYeWGY9Gq2IuDMzjx5Yvj8zD6uyJnVTKVMs+vOGdge+C9wOfDgzv1/Gawsy81/6U1ukqbxkfDgGyMy/iohJv21IGrGPTfHcP5RWhcqyZ0ScPrA8f3A5M/+0gprUQWXNQV4P/AKwDNgKPBoRj2TmoyW9fmdFxHzg14DNVdei2vvxFM89WVoV0oDMfH3VNahUfwX88sDytweWk940MGnkSp2DHBG7A6+kN83ilcC+wD2ZOdHcWc1QRPyYHS/SA/gXeh3OezPzwfKrUlNExMPAdRM9BZyRmfuXXJJERPx2Zn60//hXM/MrA899ODP/U3XVSWqrsgPyC4Dj6M1BHgvJD2fmkaUVIWlCk1zk+azM/HxZtUhjBuekTjA/dYdlNV9EXDRu1djdmL6TmRsqKEkdVdYc5I/TC8TL6F0dfxu927i8IzMfL6OGLoiIczPzj/uPX5OZtw48t9KrgDWNv6d328Vyb20jTS0meTzRsppv4QTrFgMfjIgrMnOiT7mkwpUyghwR76F3z+PvZub2kb9gRznSormIiLXAUuDveO4+5bdl5lRzk6WRsl8TQETsBfyl/79VlrIu0vsa8PhYOI6I19P74pAfAFdn5s9KqqPtHGnRrGXmsRHxb4Dj6X3i8x7gCxHxEHBrZv5mpQWqq14eEf+PXh+2oP+Y/vKu1ZWlMmXmj8LbMalEpXzVNPBlYDeAiHgF8BVgI/By4FMl1dAF479Fb7LnpAll5k/7t3r7JPBx4Bp6x+7yKutSd2XmvMzcPTMXZub8/uOx5Z2rrk/l6A+sPVZ1HeqOskaQFwzcQeFc4NrM/FhE7ETv3sgqxs9HxPfojay8pP+Y/vLS6spSE0TEOfRGjl8B/CuwBrgDODEzH6qwNHVYROwK/AZwCPA9euePbdVWpVGJiLt5/oDOXsCDwNvLr0hdVVZAHvxY5A3AZQCZ+YwfmRTq5cD+wKZx6w8GDDiazqeB++ldQPvtzPzHiuuRAD4PPA38NXAK8FLgwkor0ij9+3HLCWzNTO/FrlKVFZBviogvA/8M/BxwE0BEHAg8VVINXfBx4LLM/MHgyv79pz/Ojjdfl8bbk96brFcDV0TEYfSO2dvoXax3U4W1qbuOGLsVaER8DvjbiuvRaP2Q5z4xuBv4nJ8YqAplzUF+L71vv3mA3se1T/fXH0LvoxMVY//MvHv8yv66xeWXoybJzO2ZeWdmXp2Z59Abrfsm8C7gW9VWpw4bO19gUOqEzwPH0gvHb2LqrxqXRqaUEeT+fVWvA4iIoyLivcCvAhuAT5RRQ0fsOcVzC8oqQs0UEb9Ab/R47GcXeqPHv0fvlm9SFV4+7s4VCwbuapGZuXt1pWkE/MRAtVDWF4UcCpzd/3kU+BK9ezC/vozX75C1EXFBZn5mcGVEnE/v3rbSVP4Q+A7wDeB3MnNjteVIvbtYVF2DSrXDJwZepqSqlPVFIc/Qu8DivMxc11+3PjO9s0KBImJ/4M+An/FcID6W3kjgW7wTgaYSEW8GFo1942JE3AHs23/6tzPzq1XVJqkbImI7MHZBXtD79POn+ImBSlbWRXqnA2cBN0fEN+lNt/BtYcEy84fAq/v3i3xZf/XXvbhKQ/oAveN0zAuA4+jdB/kPAAOypJHyEwPVRVlzkK8Hro+I3YDT6F20t19E/D7wZ5n5F2XU0RWZeTNwc9V1qHF2yczBWwR+JzO3Alv7x64kSZ1QyhSLCV844ufoXah3ZmaeVEkRkp4VEesy85BJnvunzHxJ2TVJklSFsm7z9jyZ+VhmrjIcS7VxR0RcMH5lRPw6XkkuSeqQykaQJdVLROwHXE/va6bv7K8+ht5c5Df357hLktR6BmRJO4iIN9D7Ol+Ae73IU5LUNQZkSZIkaUBlc5AlSZKkOjIgS5IkSQMMyJIkSdIAA7IkSZI04P8DMFA4ZKU3rVsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%pylab inline\n",
    "\n",
    "# The code below makes ADMIXTURE bar plots based on your results\n",
    "# It will fail if your ADMIXTURE run was not complete or if you did not\n",
    "# output results to the specified output files!\n",
    "# Please ensure you get plots (and the notebook validates)\n",
    "# to get full points.\n",
    "\n",
    "sampleinfo = pd.read_csv(\"~/public/1000Genomes/igsr_samples.tsv\", sep=\"\\t\")\n",
    "sample_to_pop = dict(zip(list(sampleinfo[\"Sample name\"]), list(sampleinfo[\"Population code\"])))\n",
    "samples = [line.split()[0] for line in open(\"ps2_admixture.pruned.fam\", \"r\").readlines()]\n",
    "pops = [sample_to_pop.get(item, \"NA\") for item in samples]\n",
    "\n",
    "fig = plt.figure()\n",
    "fig.set_size_inches((10, 8))\n",
    "\n",
    "plotind = 1\n",
    "for K in [3]:\n",
    "    ax = fig.add_subplot(3, 1, plotind)\n",
    "    \n",
    "    data = pd.read_csv(\"result.Q\", delim_whitespace=True, header=None)\n",
    "    cols = list(data.columns)\n",
    "    data[\"sample\"] = samples\n",
    "    data[\"pop\"] = pops\n",
    "    data = data.sort_values([\"pop\"]+cols)\n",
    "    data.plot.bar(stacked=True, ax=ax, width=1)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom')\n",
    "    \n",
    "    # Only plot label for first sample in each pop\n",
    "    xticklabels = []\n",
    "    currpop = \"\"\n",
    "    for i in range(data.shape[0]):\n",
    "        if data[\"pop\"].values[i] == currpop:\n",
    "            xticklabels.append(\"\")\n",
    "        else:\n",
    "            xticklabels.append(data[\"pop\"].values[i])\n",
    "            currpop = data[\"pop\"].values[i]\n",
    "    ax.set_xticklabels(xticklabels)\n",
    "    \n",
    "    plotind += 1\n",
    "fig.tight_layout()\n",
    "\n",
    "\n",
    "print(\"These are the actual results from the admixture tool \" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74a5effc",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### #All the cells after this are just other codes that didn't work, you can take a look if you want but there\n",
    "##### are not considered a part of the projetct currently. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f167c599",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "#######dont run this , this is too slow. \n",
    "def naive_admixture(genotypes, num_individuals, num_snps, K=3, num_iterations=10):\n",
    "    # Initialize P (allele frequencies) randomly for each SNP for each population\n",
    "    P = np.random.rand(num_snps, K)\n",
    "    \n",
    "    # Initialize Q (ancestry proportions) randomly for each individual\n",
    "    Q = np.random.rand(num_individuals, K)\n",
    "    \n",
    "    # Normalize P and Q so they sum to 1 across the K populations\n",
    "    P /= P.sum(axis=1, keepdims=True)\n",
    "    Q /= Q.sum(axis=1, keepdims=True)\n",
    "    \n",
    "    # EM algorithm\n",
    "    for _ in range(num_iterations):\n",
    "        # E-step: Estimate ancestry fractions\n",
    "        # Here we assume genotypes are encoded as 0, 1, or 2\n",
    "        # and use a very simple model where each genotype\n",
    "        # probability is just the product of allele frequencies\n",
    "        for i in range(num_individuals):\n",
    "            for j in range(num_snps):\n",
    "                genotype = genotypes[j, i]\n",
    "                probs = P[j] ** genotype * (1 - P[j]) ** (2 - genotype)\n",
    "                if np.any(np.isnan(probs)):\n",
    "                    continue\n",
    "                Q[i] *= probs\n",
    "                \n",
    "            # Normalize Q for this individual\n",
    "            Q[i] /= Q[i].sum()\n",
    "\n",
    "        # M-step: Update allele frequencies\n",
    "        for j in range(num_snps):\n",
    "            for k in range(K):\n",
    "                # This is a very naive way to update allele frequencies,\n",
    "                # just based on the average ancestry fraction weighted by genotype\n",
    "                P[j, k] = np.sum(Q[:, k] * genotypes[j]) / (2 * np.sum(Q[:, k]))\n",
    "                \n",
    "            # Normalize P for this SNP\n",
    "            P[j] /= P[j].sum()\n",
    "\n",
    "    # Save results to files\n",
    "    np.savetxt('result.Q', Q, fmt='%f')\n",
    "    np.savetxt('result.P', P, fmt='%f')\n",
    "    \n",
    "    return Q, P\n",
    "\n",
    "# Dummy data for testing the function\n",
    "num_individuals = 668  # Replace with your actual number of individuals\n",
    "num_snps = 205857      # Replace with your actual number of SNPs\n",
    "\n",
    "# Randomly generated dummy genotype data\n",
    "# In real scenario, you would read from your .bed file\n",
    "##genotypes = np.random.randint(0, 3, (num_snps, num_individuals))\n",
    "genotypes = genotype_matrix\n",
    "# Running the naive admixture analysis\n",
    "Q, P = naive_admixture(genotypes, num_individuals, num_snps)\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bfed0c2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starting code \n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_265/292336622.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[0;31m# Read the genotype data from the .bed file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 56\u001b[0;31m \u001b[0mgenotype_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mread_bed_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbed_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_individuals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_snps\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenotype_matrix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m \u001b[0;31m# You can now proceed with your admixture analysis using the `genotype_matrix`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_265/292336622.py\u001b[0m in \u001b[0;36mread_bed_file\u001b[0;34m(bed_file_path, num_individuals, num_snps)\u001b[0m\n\u001b[1;32m     34\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mgenotype_code\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0b00\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m                     \u001b[0mgenotypes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msnp_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m  \u001b[0;31m# Homozygous first allele\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m                 \u001b[0;32melif\u001b[0m \u001b[0mgenotype_code\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0b11\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     37\u001b[0m                     \u001b[0;31m#print('hit')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m                     \u001b[0mgenotypes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msnp_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m2\u001b[0m  \u001b[0;31m# Homozygous second allele\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "## This is our code for the admixture \n",
    "\n",
    "################ code 2 for the admixtute, runs on the input bed file \n",
    "\n",
    "import numpy as np\n",
    "print('starting code ')\n",
    "def read_bed_file(bed_file_path, num_individuals, num_snps):\n",
    "    \"\"\"Read genotype data from a .bed file.\"\"\"\n",
    "    # Open the .bed file in binary mode\n",
    "    with open(bed_file_path, 'rb') as bed_file:\n",
    "        # Skip the header\n",
    "        bed_file.read(3)\n",
    "        \n",
    "        # Calculate the number of bytes to read per row\n",
    "        bytes_per_row = (num_individuals + 3) // 4\n",
    "        \n",
    "        # Initialize an empty array to store the genotype data\n",
    "        genotypes = np.zeros((num_snps, num_individuals), dtype=float)\n",
    "        \n",
    "        for snp_index in range(num_snps):\n",
    "            # Read a row of data\n",
    "            row_data = bed_file.read(bytes_per_row)\n",
    "            byte_array = np.frombuffer(row_data, dtype=np.uint8)\n",
    "            \n",
    "            # Decode the genotypes from the byte array\n",
    "            for i in range(num_individuals):\n",
    "                byte_index = i // 4\n",
    "                if byte_index >= len(byte_array):\n",
    "                    #print('breaking oh no ')\n",
    "                    break\n",
    "                bit_index = (i % 4) * 2\n",
    "                genotype_code = (byte_array[byte_index] >> bit_index) & 0b11\n",
    "                #print(genotype_code)\n",
    "                if genotype_code == 0b00:\n",
    "                    genotypes[snp_index, i] = 0  # Homozygous first allele\n",
    "                elif genotype_code == 0b11:\n",
    "                    #print('hit')\n",
    "                    genotypes[snp_index, i] = 2  # Homozygous second allele\n",
    "                elif genotype_code == 0b01  :\n",
    "                    genotypes[snp_index, i] = 1  # Heterozygous\n",
    "                else:\n",
    "                    #print( genotype_code)\n",
    "                    genotypes[snp_index, i] = np.nan  # Missing genotype\n",
    "                    \n",
    "    return genotypes\n",
    "\n",
    "# Parameters: Update these based on your .bed file\n",
    "bed_file_path = \"ps2_admixture.pruned.bed\"\n",
    "num_individuals =462   \n",
    "#668  # Update with the correct number of individuals in your dataset\n",
    "num_snps = 26499   \n",
    "##205857       # Update with the correct number of SNPs\n",
    "bytes_per_row = (num_individuals * 2 + 7) // 8\n",
    "\n",
    "# Read the genotype data from the .bed file\n",
    "genotype_matrix = read_bed_file(bed_file_path, num_individuals, num_snps)\n",
    "print(genotype_matrix.shape)\n",
    "# You can now proceed with your admixture analysis using the `genotype_matrix`\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def estimate_allele_frequencies(genotype_matrix):\n",
    "    \"\"\"Estimate allele frequencies in the genotype data.\"\"\"\n",
    "    # Count alleles (note: assumes 0, 1, 2 coding for genotypes)\n",
    "    allele_counts = np.nansum(genotype_matrix, axis=1)\n",
    "    # Count non-missing genotypes\n",
    "    non_missing_counts = np.count_nonzero(~np.isnan(genotype_matrix), axis=1) * 2\n",
    "    # Calculate frequencies\n",
    "    allele_frequencies = allele_counts / non_missing_counts\n",
    "    return allele_frequencies\n",
    "\n",
    "def estimate_admixture_proportions(genotype_matrix, ancestral_freqs):\n",
    "    \"\"\"Estimate admixture proportions based on allele frequencies.\"\"\"\n",
    "    num_individuals = genotype_matrix.shape[1]\n",
    "    admixture_proportions = np.zeros((num_individuals, len(ancestral_freqs)))\n",
    "    \n",
    "    for i, individual_genotypes in enumerate(genotype_matrix.T):  # Iterate through individuals\n",
    "        for j, freqs in enumerate(ancestral_freqs):  # Iterate through ancestral populations\n",
    "            # Calculate similarity score based on allele frequencies\n",
    "            similarity_score = np.nansum(np.abs(individual_genotypes - freqs))\n",
    "            admixture_proportions[i, j] = similarity_score\n",
    "    \n",
    "    # Normalize admixture proportions so they sum to 1 for each individual\n",
    "    admixture_proportions = 1 - (admixture_proportions / np.nansum(admixture_proportions, axis=1, keepdims=True))\n",
    "    \n",
    "    return admixture_proportions\n",
    "\n",
    "# Load your genotype_matrix from the .bed file as shown in the previous example\n",
    "\n",
    "# Hypothetical allele frequencies for two ancestral populations\n",
    "ancestral_freqs = [\n",
    "    np.random.uniform(0, 1, genotype_matrix.shape[0]),  # Ancestral pop 1\n",
    "    np.random.uniform(0, 1, genotype_matrix.shape[0])   # Ancestral pop 2\n",
    "]\n",
    "\n",
    "# Estimate admixture proportions\n",
    "admixture_proportions = estimate_admixture_proportions(genotype_matrix, ancestral_freqs)\n",
    "\n",
    "# Example: Print admixture proportions for the first 5 individuals\n",
    "print(admixture_proportions[:5])\n",
    "\n",
    "\n",
    "\n",
    "## IF we want to run it without any asssumptions about allelle frequencies.  \n",
    "\n",
    "import numpy as np\n",
    "from sklearn.cluster import KMeans\n",
    "\n",
    "def estimate_allele_frequencies(genotype_matrix):\n",
    "    \"\"\"Estimate allele frequencies in the genotype data.\"\"\"\n",
    "    # Assuming 0, 1, 2 encoding for genotypes\n",
    "    allele_counts = np.nansum(genotype_matrix, axis=1)\n",
    "    # Count non-missing genotypes\n",
    "    non_missing_counts = np.count_nonzero(~np.isnan(genotype_matrix), axis=1) * 2\n",
    "    # Calculate frequencies\n",
    "    allele_frequencies = allele_counts / non_missing_counts\n",
    "    return allele_frequencies\n",
    "\n",
    "def cluster_individuals(genotype_matrix, n_clusters=2):\n",
    "    \"\"\"Cluster individuals based on their genetic data.\"\"\"\n",
    "    # Transpose matrix to have individuals as rows\n",
    "    individuals_matrix = genotype_matrix.T\n",
    "    # Handle missing data by imputing with the mean\n",
    "    nan_mask = np.isnan(individuals_matrix)\n",
    "    mean_genotypes = np.nanmean(individuals_matrix, axis=0)\n",
    "    individuals_matrix[nan_mask] = np.take(mean_genotypes, np.where(nan_mask)[1])\n",
    "    # Use KMeans for clustering\n",
    "    kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(individuals_matrix)\n",
    "    return kmeans.labels_, kmeans.cluster_centers_\n",
    "\n",
    "def estimate_admixture_proportions(cluster_centers):\n",
    "    \"\"\"Estimate admixture proportions based on cluster centers.\"\"\"\n",
    "    # Normalize cluster centers to get admixture proportions\n",
    "    admixture_proportions = cluster_centers / np.sum(cluster_centers, axis=1, keepdims=True)\n",
    "    return admixture_proportions\n",
    "\n",
    "# Load your genotype_matrix from the .bed file as shown in the previous example\n",
    "\n",
    "# Cluster individuals (adjust n_clusters as needed)\n",
    "#labels, cluster_centers = cluster_individuals(genotype_matrix, n_clusters=2)\n",
    "\n",
    "# Estimate admixture proportions based on the cluster centers\n",
    "#admixture_proportions = estimate_admixture_proportions(cluster_centers)\n",
    "\n",
    "# Example: Print the cluster assignment for the first 5 individuals\n",
    "#print(\"Cluster labels for the first 5 individuals:\", labels[:5])\n",
    "print('done')\n",
    "dimensions = genotype_matrix.shape\n",
    "\n",
    "\n",
    "\n",
    "# Calculate frequencies of all values\n",
    "\n",
    "(unique, counts) = np.unique(genotype_matrix, return_counts=True)\n",
    "\n",
    "frequencies = np.asarray((unique, counts)).T\n",
    "\n",
    "# Change binary genotype to ped+map format\n",
    "#plink --bfile ps2_admixture.pruned.bed --cow --nonfounders --allow-no-sex --recode --out ps2_admix_trial\n",
    "\n",
    "print(dimensions, frequencies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "33c4b113",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "could not convert string to float: 'C'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_111/2621099544.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     32\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mQ\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 34\u001b[0;31m \u001b[0mP_init\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minitialize_P_from_bim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'ps2_admixture.pruned.bim'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_snps\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     35\u001b[0m \u001b[0mQ_init\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minitialize_Q_from_fam\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'ps2_admixture.pruned.fam'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_individuals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_111/2621099544.py\u001b[0m in \u001b[0;36minitialize_P_from_bim\u001b[0;34m(bim_file_path, num_snps, K)\u001b[0m\n\u001b[1;32m      7\u001b[0m             \u001b[0;31m# Otherwise, you would have to calculate it from your genotype data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m             \u001b[0;31m# Here, we're just using a placeholder value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m             \u001b[0mmaf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m             \u001b[0mallele_frequencies\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msnp_index\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mmaf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mmaf\u001b[0m\u001b[0;34m]\u001b[0m  \u001b[0;31m# Assuming the first allele is the minor allele\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: could not convert string to float: 'C'"
     ]
    }
   ],
   "source": [
    "def initialize_P_from_bim(bim_file_path, num_snps, K):\n",
    "    allele_frequencies = np.zeros((num_snps, 2))  # Assuming two alleles per SNP\n",
    "    with open(bim_file_path, 'r') as bim_file:\n",
    "        for snp_index, line in enumerate(bim_file):\n",
    "            parts = line.strip().split()\n",
    "            # parts[4] is the minor allele frequency (MAF) for this SNP, if available\n",
    "            # Otherwise, you would have to calculate it from your genotype data\n",
    "            # Here, we're just using a placeholder value\n",
    "            maf = float(parts[4])  \n",
    "            allele_frequencies[snp_index] = [maf, 1 - maf]  # Assuming the first allele is the minor allele\n",
    "    \n",
    "    # Randomly initialize P based on allele frequencies\n",
    "    # This is a simplistic way to use the allele frequencies for initialization\n",
    "    P = np.zeros((num_snps, K))\n",
    "    for k in range(K):\n",
    "        P[:, k] = allele_frequencies[:, 0] + (np.random.rand(num_snps) * (allele_frequencies[:, 1] - allele_frequencies[:, 0]))\n",
    "    \n",
    "    P /= P.sum(axis=1, keepdims=True)\n",
    "    return P\n",
    "\n",
    "def initialize_Q_from_fam(fam_file_path, num_individuals, K):\n",
    "    Q = np.zeros((num_individuals, K))\n",
    "    with open(fam_file_path, 'r') as fam_file:\n",
    "        for i, line in enumerate(fam_file):\n",
    "            parts = line.strip().split()\n",
    "            population_label = parts[5]  # Assuming population label is in the sixth column\n",
    "            # Convert population_label to an index and assign the initial ancestry proportion\n",
    "            # Here, you'd have a mapping of population labels to indices\n",
    "            Q[i, population_label_to_index[population_label]] = 1\n",
    "    \n",
    "    Q /= Q.sum(axis=1, keepdims=True)\n",
    "    return Q\n",
    "\n",
    "P_init = initialize_P_from_bim('ps2_admixture.pruned.bim', num_snps, 3)\n",
    "Q_init = initialize_Q_from_fam('ps2_admixture.pruned.fam', num_individuals, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4557a967",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Load your genotype matrix where rows are SNPs and columns are individuals\n",
    "# For demonstration, let's assume it's already loaded into a variable named `genotype_matrix`\n",
    "\n",
    "# Let's assume the genotype_matrix is coded as:\n",
    "# 0 for the homozygous reference allele\n",
    "# 1 for the heterozygous allele\n",
    "# 2 for the homozygous alternate allele\n",
    "# NaN for missing data\n",
    "\n",
    "# Calculate the frequency of the alternate allele (allele 2 from the .bim file)\n",
    "allele_2_counts = np.nansum(genotype_matrix, axis=1)  # Sum the genotypes for each SNP\n",
    "heterozygous_counts = np.sum(genotype_matrix == 1, axis=1)  # Count the heterozygotes\n",
    "allele_2_frequencies = (allele_2_counts + heterozygous_counts) / (2 * (~np.isnan(genotype_matrix)).sum(axis=1))\n",
    "\n",
    "# If you need frequencies of the reference allele, simply subtract from 1\n",
    "allele_1_frequencies = 1 - allele_2_frequencies\n",
    "import numpy as np\n",
    "\n",
    "# Assume genotype_matrix is loaded, with shape (num_snps, num_individuals)\n",
    "# and contains genotypes encoded as 0, 1, 2, and NaN for missing data.\n",
    "\n",
    "# Calculate allele 2 frequencies\n",
    "allele_2_counts = np.nansum(genotype_matrix, axis=1)  # Sum the genotypes for each SNP\n",
    "heterozygous_counts = np.sum(genotype_matrix == 1, axis=1)  # Count the heterozygotes\n",
    "allele_2_frequencies = (allele_2_counts + heterozygous_counts) / (2 * (~np.isnan(genotype_matrix)).sum(axis=1))\n",
    "\n",
    "# Initialize P using these frequencies\n",
    "num_snps, num_individuals = genotype_matrix.shape\n",
    "K = 3  # For example, assuming there are 3 populations\n",
    "P = np.zeros((num_snps, K))\n",
    "\n",
    "for k in range(K):\n",
    "    # Use the allele 2 frequencies to initialize one allele's frequency in P\n",
    "    # and assume that allele 1's frequency is (1 - allele 2 frequency)\n",
    "    # You can add some noise to these initial frequencies if desired\n",
    "    P[:, k] = allele_2_frequencies + np.random.normal(0, 0.1, num_snps)  # Adding small noise\n",
    "\n",
    "# Ensure P's frequencies are normalized\n",
    "P = np.clip(P, 0.0001, 0.9999)  # Clip values to avoid 0 or 1, which are problematic in log calculations\n",
    "P /= P.sum(axis=1, keepdims=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "49efa0b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.33496071, 0.33007857, 0.33496071],\n",
       "       [0.36782972, 0.34122868, 0.2909416 ],\n",
       "       [0.32455483, 0.33772259, 0.33772259],\n",
       "       [0.35596366, 0.34645851, 0.29757783],\n",
       "       [0.35160963, 0.31503154, 0.33335884],\n",
       "       [0.32718228, 0.3298621 , 0.34295562],\n",
       "       [0.33333333, 0.33333333, 0.33333333],\n",
       "       [0.34227167, 0.31545666, 0.34227167],\n",
       "       [0.31993438, 0.35651119, 0.32355442],\n",
       "       [0.37649258, 0.29033802, 0.3331694 ],\n",
       "       [0.35269232, 0.31124179, 0.33606588],\n",
       "       [0.32994447, 0.34000176, 0.33005377],\n",
       "       [0.34934535, 0.29839873, 0.35225592],\n",
       "       [0.34657085, 0.32702492, 0.32640424],\n",
       "       [0.34026979, 0.37989269, 0.27983753],\n",
       "       [0.34266054, 0.34816223, 0.30917723],\n",
       "       [0.33283235, 0.24873967, 0.41842797],\n",
       "       [0.35544937, 0.33131109, 0.31323954],\n",
       "       [0.30764486, 0.31823334, 0.37412179],\n",
       "       [0.31457598, 0.34271201, 0.34271201],\n",
       "       [0.30355408, 0.34822296, 0.34822296],\n",
       "       [0.29473722, 0.35006612, 0.35519666],\n",
       "       [0.31560736, 0.34093136, 0.34346128],\n",
       "       [0.31371442, 0.33788621, 0.34839937],\n",
       "       [0.32761417, 0.34675736, 0.32562847],\n",
       "       [0.33333333, 0.33333333, 0.33333333],\n",
       "       [0.34434726, 0.29712889, 0.35852385],\n",
       "       [0.32890355, 0.37270417, 0.29839228],\n",
       "       [0.31519611, 0.34158933, 0.34321457],\n",
       "       [0.300037  , 0.3499815 , 0.3499815 ],\n",
       "       [0.34349237, 0.29676762, 0.35974   ],\n",
       "       [0.3514913 , 0.32131508, 0.32719362],\n",
       "       [0.25513139, 0.40266497, 0.34220364],\n",
       "       [0.34629481, 0.34629481, 0.30741039],\n",
       "       [0.28749245, 0.3302821 , 0.38222545],\n",
       "       [0.31467289, 0.34575805, 0.33956906],\n",
       "       [0.31729824, 0.32289987, 0.35980189],\n",
       "       [0.30566711, 0.34726581, 0.34706708],\n",
       "       [0.33383372, 0.33383372, 0.33233255],\n",
       "       [0.36369576, 0.31139371, 0.32491054],\n",
       "       [0.34497016, 0.31059344, 0.3444364 ],\n",
       "       [0.32970688, 0.33514656, 0.33514656],\n",
       "       [0.34734365, 0.29517881, 0.35747755],\n",
       "       [0.39817846, 0.3015067 , 0.30031484],\n",
       "       [0.35558732, 0.30584578, 0.3385669 ],\n",
       "       [0.34846693, 0.34846693, 0.30306614],\n",
       "       [0.33959718, 0.343969  , 0.31643382],\n",
       "       [0.33568383, 0.33111429, 0.33320188],\n",
       "       [0.35740684, 0.30860795, 0.33398521],\n",
       "       [0.35288083, 0.37076824, 0.27635093],\n",
       "       [0.39632303, 0.31631527, 0.2873617 ],\n",
       "       [0.3282556 , 0.31128091, 0.36046349],\n",
       "       [0.33382921, 0.33234157, 0.33382921],\n",
       "       [0.30372263, 0.33703118, 0.35924619],\n",
       "       [0.3523489 , 0.30444662, 0.34320449],\n",
       "       [0.32904514, 0.3295906 , 0.34136426],\n",
       "       [0.33333333, 0.33333333, 0.33333333],\n",
       "       [0.38162202, 0.32102774, 0.29735024],\n",
       "       [0.3544416 , 0.28851948, 0.35703892],\n",
       "       [0.33337545, 0.27658825, 0.39003631],\n",
       "       [0.36004092, 0.32746195, 0.31249713],\n",
       "       [0.29973671, 0.35165996, 0.34860333],\n",
       "       [0.34768974, 0.33363253, 0.31867773],\n",
       "       [0.32767395, 0.35468706, 0.31763899],\n",
       "       [0.34480836, 0.32149715, 0.33369449],\n",
       "       [0.29922286, 0.36406595, 0.33671119],\n",
       "       [0.34801916, 0.31602841, 0.33595243],\n",
       "       [0.32850237, 0.33286392, 0.33863371],\n",
       "       [0.32685884, 0.34638102, 0.32676013],\n",
       "       [0.35822132, 0.34581629, 0.29596239],\n",
       "       [0.35752892, 0.33945259, 0.30301848],\n",
       "       [0.30607201, 0.35904743, 0.33488056],\n",
       "       [0.31674297, 0.35305532, 0.33020171],\n",
       "       [0.32652549, 0.33673725, 0.33673725],\n",
       "       [0.34056662, 0.30394944, 0.35548394],\n",
       "       [0.34967117, 0.34967117, 0.30065765],\n",
       "       [0.32870028, 0.35041185, 0.32088787],\n",
       "       [0.35082783, 0.35194696, 0.29722521],\n",
       "       [0.31474479, 0.33691512, 0.34834009],\n",
       "       [0.33443762, 0.33443762, 0.33112476],\n",
       "       [0.33333333, 0.33333333, 0.33333333],\n",
       "       [0.34695579, 0.34205373, 0.31099048],\n",
       "       [0.34087154, 0.35211027, 0.3070182 ],\n",
       "       [0.35744626, 0.31929918, 0.32325456],\n",
       "       [0.34751495, 0.30497009, 0.34751495],\n",
       "       [0.35208522, 0.3403017 , 0.30761308],\n",
       "       [0.34054543, 0.34332089, 0.31613368],\n",
       "       [0.31903569, 0.32022895, 0.36073535],\n",
       "       [0.35961848, 0.28106927, 0.35931226],\n",
       "       [0.35589963, 0.31606058, 0.32803979],\n",
       "       [0.29267461, 0.33517772, 0.37214767],\n",
       "       [0.34787976, 0.30424048, 0.34787976],\n",
       "       [0.27818062, 0.34078792, 0.38103146],\n",
       "       [0.36023652, 0.30002788, 0.3397356 ],\n",
       "       [0.32056639, 0.34920567, 0.33022794],\n",
       "       [0.29652112, 0.32056488, 0.382914  ],\n",
       "       [0.34412804, 0.32514354, 0.33072842],\n",
       "       [0.30457857, 0.36199041, 0.33343102],\n",
       "       [0.33333333, 0.33333333, 0.33333333],\n",
       "       [0.34068647, 0.34068647, 0.31862706]])"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_2_frequencies.shape\n",
    "P[0:100]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23616101",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26499, 462)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_matrix[:10]\n",
    "genotype_matrix.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cd7b67c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting the complex algo with Quasi Newton involved \n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy.optimize import minimize\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "print(\"Starting the complex algo with Quasi Newton involved \")\n",
    "# Assuming `genotype_matrix`, `num_individuals`, and `num_snps` are defined\n",
    "# targ_Q and targ_P represent the target matrices for comparison\n",
    "\n",
    "# Define the negative log-likelihood function\n",
    "def neg_log_likelihood(params, genotypes, num_individuals, num_snps, K):\n",
    "    P = params[:num_snps * K].reshape((num_snps, K))\n",
    "    Q = params[num_snps * K:].reshape((num_individuals, K))\n",
    "    nll = 0.0\n",
    "    for i in range(num_individuals):\n",
    "        for j in range(num_snps):\n",
    "            genotype = genotypes[j, i]\n",
    "            if not np.isnan(genotype):\n",
    "                # Calculate the probability of the genotype given P and Q\n",
    "                prob_geno = 0.0\n",
    "                for k in range(K):\n",
    "                    # Assuming Hardy-Weinberg Equilibrium for the allele frequencies\n",
    "                    if genotype == 2.0:  # Homozygous for allele 2\n",
    "                        prob_geno += Q[i, k] * (P[j, k] ** 2)\n",
    "                    elif genotype == 1.0:  # Heterozygous\n",
    "                        prob_geno += 2 * Q[i, k] * P[j, k] * (1 - P[j, k])\n",
    "                    elif genotype == 0.0:  # Homozygous for allele 1\n",
    "                        prob_geno += Q[i, k] * ((1 - P[j, k]) ** 2)\n",
    "                nll -= np.log(prob_geno + 1e-10)  # Add a small value to avoid log(0)\n",
    "    return nll\n",
    "\n",
    "\n",
    "def gradient(params, genotypes, num_individuals, num_snps, K):\n",
    "    P = params[:num_snps * K].reshape((num_snps, K))\n",
    "    Q = params[num_snps * K:].reshape((num_individuals, K))\n",
    "    \n",
    "    # Placeholder for the actual gradient computation\n",
    "    # The gradients should be the same shape as the parameter array\n",
    "    grad_P = np.zeros_like(P)\n",
    "    grad_Q = np.zeros_like(Q)\n",
    "    \n",
    "    # Compute the gradient here...\n",
    "    # This requires calculating the partial derivatives of the negative log-likelihood\n",
    "    # with respect to each element of P and Q\n",
    "    calculate_gradient_P(Q, genotypes, P, num_individuals, num_snps)\n",
    "    calculate_gradient_Q(P, genotypes, Q, num_individuals, num_snps)\n",
    "    \n",
    "    # Flatten and concatenate gradients to match the shape of 'params'\n",
    "    return np.concatenate([grad_P.flatten(), grad_Q.flatten()])\n",
    "def calculate_gradient_P(Q, genotypes, P, individual, snp):\n",
    "    \"\"\"\n",
    "    Calculate the gradient of the log-likelihood with respect to allele frequencies P.\n",
    "    \"\"\"\n",
    "    grad_P = 0.0\n",
    "    genotype = genotypes[snp-1, individual-1]\n",
    "    \n",
    "    if not np.isnan(genotype):\n",
    "        # Probability of the genotype under Hardy-Weinberg Equilibrium\n",
    "        prob_geno = (P[snp-1] ** 2) * (genotype == 2) + \\\n",
    "                    (2 * P[snp-1] * (1 - P[snp-1])) * (genotype == 1) + \\\n",
    "                    ((1 - P[snp-1]) ** 2) * (genotype == 0)\n",
    "        \n",
    "        # Derivative with respect to P\n",
    "        if genotype == 2:\n",
    "            grad_P = 2 * Q[individual-1] * P[snp-1] / prob_geno\n",
    "        elif genotype == 1:\n",
    "            grad_P = Q[individual-1] * (1 - 2 * P[snp-1]) / prob_geno\n",
    "        elif genotype == 0:\n",
    "            grad_P = -2 * Q[individual-1] * (1 - P[snp-1]) / prob_geno\n",
    "    \n",
    "    return grad_P\n",
    "\n",
    "def calculate_gradient_Q(P, genotypes, Q, individual, snp):\n",
    "    \"\"\"\n",
    "    Calculate the gradient of the log-likelihood with respect to ancestry proportions Q.\n",
    "    \"\"\"\n",
    "    grad_Q = 0.0\n",
    "    genotype = genotypes[snp-1, individual-1]\n",
    "    \n",
    "    if not np.isnan(genotype):\n",
    "        # Probability of the genotype under Hardy-Weinberg Equilibrium\n",
    "        prob_geno = (P[snp-1] ** 2) * (genotype == 2) + \\\n",
    "                    (2 * P[snp-1] * (1 - P[snp-1])) * (genotype == 1) + \\\n",
    "                    ((1 - P[snp-1]) ** 2) * (genotype == 0)\n",
    "        \n",
    "        # Derivative with respect to Q\n",
    "        grad_Q = np.log(prob_geno) * (genotype == 2) + \\\n",
    "                 np.log(1 - prob_geno) * (genotype == 0)\n",
    "    \n",
    "    return grad_Q\n",
    "\n",
    "\n",
    "def naive_admixture_with_quasi_newton(genotypes, num_individuals, num_snps, K, num_iterations=2500):\n",
    "    # Flatten the initial P and Q matrices for use with BFGS\n",
    "    initial_P = (np.random.rand(num_snps, K) / 5) + 0.4\n",
    "    initial_P /= initial_P.sum(axis=1, keepdims=True)\n",
    "    initial_Q = (np.random.rand(num_individuals, K) / 5) + 0.4\n",
    "    initial_Q /= initial_Q.sum(axis=1, keepdims=True)\n",
    "    initial_params = np.hstack((initial_P.flatten(), initial_Q.flatten()))\n",
    "    \n",
    "    # Run the BFGS optimization\n",
    "    result = minimize(fun=neg_log_likelihood,\n",
    "                      x0=initial_params,\n",
    "                      args=(genotypes, num_individuals, num_snps, K),\n",
    "                      jac=gradient,  # If you provide the gradient\n",
    "                      method='BFGS',\n",
    "                      options={'disp': True, 'maxiter': num_iterations})\n",
    "    \n",
    "    # Extract the optimized P and Q matrices\n",
    "    optimized_params = result.x\n",
    "    P = optimized_params[:num_snps * K].reshape((num_snps, K))\n",
    "    Q = optimized_params[num_snps * K:].reshape((num_individuals, K))\n",
    "    \n",
    "    if iter % 2 == 0:\n",
    "            l1_P = np.abs((targ_P - P)).mean().round(3)\n",
    "            l1_Q = np.abs((targ_Q - Q)).mean().round(3)\n",
    "            print('P L1 loss = {}\\nQ L1 Loss = {}'.format(l1_P, l1_Q))\n",
    "    \n",
    "    return P, Q\n",
    "\n",
    "# Run the admixture analysis with quasi-Newton acceleration\n",
    "P, Q = naive_admixture_with_quasi_newton(genotype_matrix, num_individuals, num_snps, K=3, num_iterations=2500)\n",
    "print('Optimization done')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da2f97cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a6b5bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc561488",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4fd33d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "a112fa15",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Load the .raw file generated by PLINK\n",
    "raw_data = pd.read_csv('orig.raw', delim_whitespace=True)\n",
    "\n",
    "# Drop non-genotype columns. Assuming the first six columns are non-genotype data\n",
    "genotype_data = raw_data.drop(raw_data.columns[:6], axis=1)\n",
    "\n",
    "# Transpose the dataframe so that SNPs are rows and individuals are columns\n",
    "genotype_matrix = genotype_data.transpose()\n",
    "\n",
    "# Replace 0 with NaN to indicate missing data\n",
    "genotype_matrix.replace(0, np.nan, inplace=True)\n",
    "\n",
    "# Save to a .csv file\n",
    "genotype_matrix.to_csv('genotype_matrix.csv', header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b5a17166",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>452</th>\n",
       "      <th>453</th>\n",
       "      <th>454</th>\n",
       "      <th>455</th>\n",
       "      <th>456</th>\n",
       "      <th>457</th>\n",
       "      <th>458</th>\n",
       "      <th>459</th>\n",
       "      <th>460</th>\n",
       "      <th>461</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs539713234_C</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs78200054_T</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs559733149_T</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs564500555_G</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs71235073_T</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs368646645_C</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs71235074_G</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs71235075_T</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs533016209_A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs373407518_T</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 462 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0    1    2    3    4    5    6    7    8    9    ...  452  \\\n",
       "rs539713234_C  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN   \n",
       "rs78200054_T   1.0  1.0  1.0  NaN  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0   \n",
       "rs559733149_T  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN   \n",
       "rs564500555_G  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN   \n",
       "rs71235073_T   1.0  1.0  1.0  NaN  NaN  1.0  1.0  1.0  1.0  1.0  ...  1.0   \n",
       "rs368646645_C  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  NaN  1.0  ...  1.0   \n",
       "rs71235074_G   1.0  NaN  1.0  1.0  1.0  1.0  NaN  1.0  1.0  1.0  ...  1.0   \n",
       "rs71235075_T   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0   \n",
       "rs533016209_A  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN   \n",
       "rs373407518_T  1.0  1.0  1.0  NaN  NaN  1.0  1.0  1.0  1.0  1.0  ...  1.0   \n",
       "\n",
       "               453  454  455  456  457  458  459  460  461  \n",
       "rs539713234_C  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "rs78200054_T   1.0  1.0  1.0  1.0  1.0  1.0  1.0  2.0  1.0  \n",
       "rs559733149_T  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  \n",
       "rs564500555_G  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "rs71235073_T   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  \n",
       "rs368646645_C  1.0  NaN  1.0  1.0  1.0  1.0  NaN  NaN  NaN  \n",
       "rs71235074_G   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  \n",
       "rs71235075_T   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  \n",
       "rs533016209_A  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  \n",
       "rs373407518_T  NaN  1.0  2.0  1.0  1.0  1.0  1.0  NaN  1.0  \n",
       "\n",
       "[10 rows x 462 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_matrix[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "054ee360",
   "metadata": {},
   "outputs": [],
   "source": [
    "##plink --bfile ps2_admixture --recode A-transpose --out orig\n",
    "##plink --bfile ps2_admixture --recode A --transpose --out orig"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
